

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Benchmarking Examples &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Benchmarking Examples</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/benchmarking_examples.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="benchmarking-examples">
<span id="id1"></span><h1>Benchmarking Examples<a class="headerlink" href="#benchmarking-examples" title="Link to this heading"></a></h1>
<p>Following is the code for the examples used in the benchmarking. The scripts will usually work both as a command line script,
but also includes a test showing how to use it in python.</p>
</section>
<section id="bam-filtering">
<h1>BAM Filtering<a class="headerlink" href="#bam-filtering" title="Link to this heading"></a></h1>
<p>Read a BAM file and filter out reads with MAPQ &lt; 60:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">chunk</span><span class="p">[</span><span class="n">chunk</span><span class="o">.</span><span class="n">mapq</span> <span class="o">==</span> <span class="mi">60</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="unique-intersection">
<h1>Unique Intersection<a class="headerlink" href="#unique-intersection" title="Link to this heading"></a></h1>
<p>Filter one bed file on whether the intervals overlap with another bed file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">time</span>

<span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>


<span class="k">def</span> <span class="nf">unique_intersect</span><span class="p">(</span><span class="n">filename_a</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filename_b</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chrom_sizes_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
    <span class="c1"># Load the genome</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">Genome</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">chrom_sizes_file</span><span class="p">,</span> <span class="n">filter_function</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

    <span class="c1"># Load the intervals from the second file and get the binary mask</span>
    <span class="n">genome_mask</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_intervals</span><span class="p">(</span><span class="n">filename_b</span><span class="p">)</span><span class="o">.</span><span class="n">get_mask</span><span class="p">()</span>

    <span class="k">with</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">out_filename</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">intervals</span> <span class="ow">in</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">filename_a</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="n">genome_mask</span><span class="p">[</span><span class="n">intervals</span><span class="p">]</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">intervals</span><span class="p">[</span><span class="n">mask</span><span class="p">])</span>


<span class="k">def</span> <span class="nf">test</span><span class="p">():</span>
    <span class="n">out_filename</span> <span class="o">=</span> <span class="s1">&#39;example_data/ctcf_intersect.bed.gz&#39;</span>
    <span class="n">unique_intersect</span><span class="p">(</span><span class="s1">&#39;example_data/ctcf.bed.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;example_data/znf263.bed.gz&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;example_data/hg38.chrom.sizes&#39;</span><span class="p">,</span>
                     <span class="n">out_filename</span><span class="p">)</span>

    <span class="k">assert</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="n">out_filename</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3951</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">t</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">unique_intersect</span><span class="p">(</span><span class="o">*</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Time:&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">t</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="jaccard-index">
<h1>Jaccard Index<a class="headerlink" href="#jaccard-index" title="Link to this heading"></a></h1>
<p>Calculate the Jaccard index between two or more bed files:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Calculates the jaccard similarity between all pairs of bed files.</span>
<span class="sd">Writes the results to stdout.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Dict</span>
<span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>



<span class="k">def</span> <span class="nf">jaccard_func</span><span class="p">(</span><span class="n">mask_a</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">genomic_data</span><span class="o">.</span><span class="n">GenomicArray</span><span class="p">,</span> <span class="n">mask_b</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">genomic_data</span><span class="o">.</span><span class="n">GenomicArray</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;Jaccard = intersection / union&#39;&#39;&#39;</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">mask_a</span> <span class="o">&amp;</span> <span class="n">mask_b</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="n">mask_a</span> <span class="o">|</span> <span class="n">mask_b</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">jaccard</span><span class="p">(</span><span class="n">chrom_sizes_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">bed_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="nb">float</span><span class="p">]:</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">Genome</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">chrom_sizes_file</span><span class="p">)</span>
    <span class="n">masks</span> <span class="o">=</span> <span class="p">{</span><span class="n">filename</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_intervals</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span><span class="o">.</span><span class="n">get_mask</span><span class="p">()</span> <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">bed_files</span><span class="p">}</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{(</span><span class="n">file_a</span><span class="p">,</span> <span class="n">file_b</span><span class="p">):</span> <span class="n">jaccard_func</span><span class="p">(</span><span class="n">masks</span><span class="p">[</span><span class="n">file_a</span><span class="p">],</span> <span class="n">masks</span><span class="p">[</span><span class="n">file_b</span><span class="p">])</span>
               <span class="k">for</span> <span class="n">file_a</span><span class="p">,</span> <span class="n">file_b</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">masks</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">r</span><span class="o">=</span><span class="mi">2</span><span class="p">)}</span>
    <span class="k">return</span> <span class="n">results</span>


<span class="k">def</span> <span class="nf">test</span><span class="p">():</span>
    <span class="n">chrom_sizes</span> <span class="o">=</span> <span class="s1">&#39;example_data/hg38.chrom.sizes&#39;</span>
    <span class="n">files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;example_data/ctcf.bed.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;example_data/znf263.bed.gz&#39;</span><span class="p">]</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">jaccard</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="p">,</span> <span class="n">files</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">chrom_sizes</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">files</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">jaccard</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="p">,</span> <span class="n">files</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="count-kmers">
<h1>Count Kmers<a class="headerlink" href="#count-kmers" title="Link to this heading"></a></h1>
<p>Count the number of kmers in a fasta/fastq file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>


<span class="k">def</span> <span class="nf">count_kmers</span><span class="p">(</span><span class="n">sequence_entries</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">EncodedArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">bnp</span><span class="o">.</span><span class="n">EncodedCounts</span><span class="p">:</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">as_encoded_array</span><span class="p">(</span><span class="n">sequence_entries</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">)</span>
    <span class="n">kmers</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">get_kmers</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_encoded</span><span class="p">(</span><span class="n">kmers</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">count_all_kmers</span><span class="p">(</span><span class="n">input_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
    <span class="n">buffer_type</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">TwoLineFastaBuffer</span> <span class="k">if</span> <span class="n">input_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">((</span><span class="s1">&#39;.fa&#39;</span><span class="p">,</span> <span class="s1">&#39;.fa.gz&#39;</span><span class="p">))</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">stream</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">input_file</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">buffer_type</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
    <span class="n">kmers</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">count_kmers</span><span class="p">(</span><span class="n">chunk</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span> <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">stream</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">kmer</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span>
                     <span class="k">for</span> <span class="n">kmer</span><span class="p">,</span> <span class="n">count</span>
                     <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">kmers</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span> <span class="n">kmers</span><span class="o">.</span><span class="n">counts</span><span class="p">)))</span>


<span class="k">def</span> <span class="nf">test</span><span class="p">():</span>
    <span class="n">count_all_kmers</span><span class="p">(</span><span class="s1">&#39;example_data/big.fq.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;tmp.csv&#39;</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">count_all_kmers</span><span class="p">(</span><span class="o">*</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
</pre></div>
</div>
</section>
<section id="reverse-complement">
<h1>Reverse Complement<a class="headerlink" href="#reverse-complement" title="Link to this heading"></a></h1>
<p>Reverse complement a fasta/fastq file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>


<span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="n">input_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Reverse complements a fasta or fastq file and writes the result to a new file.&quot;&quot;&quot;</span>
    <span class="n">bt</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">filename</span><span class="p">:</span> <span class="p">(</span><span class="n">bnp</span><span class="o">.</span><span class="n">TwoLineFastaBuffer</span> <span class="k">if</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">((</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="s1">&#39;fa.gz&#39;</span><span class="p">))</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">output_filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">bt</span><span class="p">(</span><span class="n">output_filename</span><span class="p">))</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">input_filename</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">bt</span><span class="p">(</span><span class="n">input_filename</span><span class="p">))</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
            <span class="n">rc</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">get_reverse_complement</span><span class="p">(</span><span class="n">chunk</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bnp</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">chunk</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">rc</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">test</span><span class="p">():</span>
    <span class="n">reverse_complement</span><span class="p">(</span><span class="s1">&#39;example_data/big.fq.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;example_data/big_rc.fq.gz&#39;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="s1">&#39;example_data/big_rc.fq.gz&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="s1">&#39;example_data/big.fq.gz&#39;</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">reverse_complement</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="sequence-length-distribution">
<h1>Sequence Length Distribution<a class="headerlink" href="#sequence-length-distribution" title="Link to this heading"></a></h1>
<p>Calculate the length distribution of sequences in a fasta/fastq file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<span class="k">def</span> <span class="nf">get_sorted_sequence_length_distribution</span><span class="p">(</span><span class="n">file_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;Return the sequence lengths found and their frequency&#39;&#39;&#39;</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">chunk</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">lengths</span> <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">())</span>
    <span class="n">sizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flatnonzero</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sizes</span><span class="p">,</span> <span class="n">counts</span><span class="p">[</span><span class="n">sizes</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">test</span><span class="p">():</span>
    <span class="n">get_sorted_sequence_length_distribution</span><span class="p">(</span><span class="s2">&quot;example_data/big.fq.gz&quot;</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">file_name</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">out_file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">sizes</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">get_sorted_sequence_length_distribution</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2"> </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">size</span><span class="p">,</span> <span class="n">count</span><span class="p">)</span> <span class="k">for</span> <span class="n">size</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">sizes</span><span class="p">,</span> <span class="n">counts</span><span class="p">)])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="subsampling">
<h1>Subsampling<a class="headerlink" href="#subsampling" title="Link to this heading"></a></h1>
<p>Subsample exactly half of the seqeuences in a fasta/fastq file. This example is complicated when working on big files as it
one have to figure out how many sequences to subsample from each chunk to get exactly half of the sequences:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dataclasses</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">islice</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>


<span class="nd">@dataclasses</span><span class="o">.</span><span class="n">dataclass</span>
<span class="k">class</span> <span class="nc">State</span><span class="p">:</span>
    <span class="n">n_entries_left</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">n_samples_left</span><span class="p">:</span> <span class="nb">int</span>


<span class="k">def</span> <span class="nf">subsample_reads</span><span class="p">(</span><span class="n">input_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
    <span class="n">bt</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">TwoLineFastaBuffer</span> <span class="k">if</span> <span class="n">input_filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">((</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;fa.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;fasta.gz&#39;</span><span class="p">))</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">n_entries</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="n">input_filename</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">bt</span><span class="p">)</span>
    <span class="n">state</span> <span class="o">=</span> <span class="n">State</span><span class="p">(</span><span class="n">n_entries_left</span><span class="o">=</span><span class="n">n_entries</span><span class="p">,</span>
                  <span class="n">n_samples_left</span><span class="o">=</span><span class="n">n_entries</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">subsets</span> <span class="o">=</span> <span class="p">(</span><span class="n">sample_from_chunk</span><span class="p">(</span><span class="n">chunk</span><span class="p">,</span> <span class="n">state</span><span class="p">)</span> <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">input_filename</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">bt</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">())</span>
    <span class="k">with</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">output_filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">bt</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">subset</span> <span class="ow">in</span> <span class="n">islice</span><span class="p">(</span><span class="n">subsets</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span>
            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">subset</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">sample_from_chunk</span><span class="p">(</span><span class="n">chunk</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">SequenceEntry</span><span class="p">,</span> <span class="n">state</span><span class="p">:</span> <span class="n">State</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">bnp</span><span class="o">.</span><span class="n">SequenceEntry</span><span class="p">:</span>
    <span class="n">chunk_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
    <span class="n">n_samples_in_this_chunk</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">hypergeometric</span><span class="p">(</span><span class="n">chunk_size</span><span class="p">,</span> <span class="n">state</span><span class="o">.</span><span class="n">n_entries_left</span> <span class="o">-</span> <span class="n">chunk_size</span><span class="p">,</span>
                                                       <span class="n">state</span><span class="o">.</span><span class="n">n_samples_left</span><span class="p">)</span>
    <span class="n">rows</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">chunk_size</span><span class="p">),</span> <span class="n">n_samples_in_this_chunk</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">subset</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">rows</span><span class="p">)]</span>
    <span class="n">state</span><span class="o">.</span><span class="n">n_samples_left</span> <span class="o">-=</span> <span class="n">n_samples_in_this_chunk</span>
    <span class="n">state</span><span class="o">.</span><span class="n">n_entries_left</span> <span class="o">-=</span> <span class="n">chunk_size</span>
    <span class="k">return</span> <span class="n">subset</span>


<span class="k">def</span> <span class="nf">test</span><span class="p">():</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="s1">&#39;example_data/big.fq.gz&#39;</span>
    <span class="n">out_filename</span> <span class="o">=</span> <span class="s1">&#39;tmp.fq.gz&#39;</span>
    <span class="n">subsample_reads</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">out_filename</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="n">out_filename</span><span class="p">)</span> <span class="o">==</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">subsample_reads</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="translation">
<h1>Translation<a class="headerlink" href="#translation" title="Link to this heading"></a></h1>
<p>Translate DNA sequences in a fasta file to protein sequences:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="kn">from</span> <span class="nn">bionumpy.sequence</span> <span class="kn">import</span> <span class="n">translate_dna_to_protein</span>


<span class="k">def</span> <span class="nf">translate_dna</span><span class="p">(</span><span class="n">input_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
    <span class="k">with</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
            <span class="n">translated</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">chunk</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">translate_dna_to_protein</span><span class="p">(</span><span class="n">chunk</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">translated</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">test_translate_dna</span><span class="p">():</span>
    <span class="n">input_file</span> <span class="o">=</span> <span class="s2">&quot;example_data/dna_translatable.fa&quot;</span>
    <span class="n">output_file</span> <span class="o">=</span> <span class="s2">&quot;tmp.fa&quot;</span>
    <span class="n">translate_dna</span><span class="p">(</span><span class="n">input_file</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span> <span class="o">==</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">translate_dna</span><span class="p">(</span><span class="o">*</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
</pre></div>
</div>
</section>
<section id="vcf-filtering">
<h1>VCF Filtering<a class="headerlink" href="#vcf-filtering" title="Link to this heading"></a></h1>
<p>Filter a VCF file on the allele count:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>


<span class="k">def</span> <span class="nf">filter_file_on_allele_count</span><span class="p">(</span><span class="n">input_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">min_ac</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
    <span class="k">with</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
            <span class="n">output_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">chunk</span><span class="p">[</span><span class="n">chunk</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">AC</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="n">min_ac</span><span class="p">])</span>


<span class="k">def</span> <span class="nf">test</span><span class="p">():</span>
    <span class="n">filter_file_on_allele_count</span><span class="p">(</span><span class="s2">&quot;example_data/variants_with_header.vcf&quot;</span><span class="p">,</span> <span class="s2">&quot;test.vcf&quot;</span><span class="p">,</span> <span class="n">min_ac</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_entries</span><span class="p">(</span><span class="s2">&quot;test.vcf&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">53</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>

    <span class="n">filter_file_on_allele_count</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">]))</span>
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>