

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Extracting kmers around snps &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Numpy Interoperability" href="../interoperability/numpy_operations.html" />
    <link rel="prev" title="Computing the similarity between two BED-files" href="similarity_measures.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Extracting kmers around snps</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/extracting_kmers_around_snps.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="extracting-kmers-around-snps">
<h1>Extracting kmers around snps<a class="headerlink" href="#extracting-kmers-around-snps" title="Link to this heading"></a></h1>
<p>This is a small example of how you can use BioNumPy to extract kmers around SNPs.</p>
<p>We have a set of SNPs in a vcf-file and a reference genome, and for every SNP we want to extract kmers
that “overlap” with that SNP both for the reference allele and the variant allele.</p>
<p>For instance, assume we have a reference genome <cite>AAAATAAA</cite> and a SNP <cite>T/G</cite> at position 4.</p>
<p>Then BioNumPy lets us crate windows around this SNP and extract sequences from those windows. If <cite>k=3</cite>, we would expect to get the reference kmers
<cite>AAA</cite>, <cite>ATA</cite> and <cite>TAA</cite> and the “variant allele” kmers kmers <cite>AAG</cite>, <cite>AGA</cite>, <cite>GAA</cite>.</p>
<p>The following is code for doing this. First we read the reference genome and variants, and filter so that we only have SNPs and not indels:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="n">k</span> <span class="o">=</span> <span class="mi">5</span>

<span class="c1"># Read genome and variants</span>
<span class="n">genome</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">Genome</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="s2">&quot;example_data/sacCer3.fa&quot;</span><span class="p">,</span> <span class="n">filter_function</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_locations</span><span class="p">(</span><span class="s2">&quot;example_data/sacCer3_sample_variants.vcf.gz&quot;</span><span class="p">,</span> <span class="n">has_numeric_chromosomes</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># get only snps from vcf</span>
<span class="n">snp_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">get_data_field</span><span class="p">(</span><span class="s1">&#39;ref_seq&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">get_data_field</span><span class="p">(</span><span class="s1">&#39;alt_seq&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span><span class="n">snp_mask</span><span class="p">]</span>
</pre></div>
</div>
<p>We  then extract windows around the variants and reads sequences from the reference genome in these windows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get windows around these snps</span>
<span class="n">windows</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">get_windows</span><span class="p">(</span><span class="n">flank</span><span class="o">=</span><span class="n">k</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Use the windows to extract sequences (kmers)</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_sequence</span><span class="p">()[</span><span class="n">windows</span><span class="p">]</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">as_encoded_array</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">)</span>
<span class="n">reference_kmers</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">get_kmers</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
</pre></div>
</div>
<p>Now <cite>reference_kmers</cite> is a RaggedArray where each row contains kmers for a variant. For instance, the reference kmers for the first three variants are:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">reference_kmers</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[TTACC, TACCC, ACCCA, CCCAT, CCATA]
[ATAAC, TAACG, AACGC, ACGCC, CGCCC]
[TTTTG, TTTGA, TTGAT, TGATA, GATAT]
</pre></div>
</div>
<p>For getting kmers from the variant alleles, we replace the allele bases with the variant allele bases:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract kmers with alternative allele on SNPs</span>
<span class="n">sequences</span><span class="p">[:,</span> <span class="n">k</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">get_data_field</span><span class="p">(</span><span class="s1">&#39;alt_seq&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">as_encoded_array</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">)</span>
<span class="n">alt_kmers</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">get_kmers</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">alt_kmers</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
</pre></div>
</div>
<p>The <cite>alt_kmers</cite> now contains kmers for the alt alleles for each variant.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[TTACG, TACGC, ACGCA, CGCAT, GCATA]
[ATAAA, TAAAG, AAAGC, AAGCC, AGCCC]
[TTTTA, TTTAA, TTAAT, TAATA, AATAT]
</pre></div>
</div>
<p>Note that these kmers are encoded using a <cite>KmerEncoding</cite>. We can get the raw encoded kmers (int64) values:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">raw_kmers</span> <span class="o">=</span> <span class="n">alt_kmers</span><span class="o">.</span><span class="n">raw</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">raw_kmers</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[591 403 100 793 198]
[ 12 515 384 352 344]
[255  63 783 195 816]
[ 66 528 132 801 712]
[912 228 825 718 435]
</pre></div>
</div>
<p>We can also call <cite>.ravel()</cite> on the RaggedArray to get one flat array with all the kmers, and write these to file.
We use <cite>to_string()</cite> on each kmer to get the string kmer. Note that this works well on small datasets, but for large
amounts of kmers, iterating each kmer and calling <cite>to_string()</cite> will be slow:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;my_kmers.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="c1"># write the first 100 kmers to file</span>
    <span class="n">kmers_as_strings</span> <span class="o">=</span> <span class="p">(</span><span class="n">kmer</span><span class="o">.</span><span class="n">to_string</span><span class="p">()</span> <span class="k">for</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="n">reference_kmers</span><span class="o">.</span><span class="n">ravel</span><span class="p">()[</span><span class="mi">0</span><span class="p">:</span><span class="mi">100</span><span class="p">])</span>
    <span class="n">f</span><span class="o">.</span><span class="n">writelines</span><span class="p">((</span><span class="n">k</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">kmers_as_strings</span><span class="p">))</span>
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="similarity_measures.html" class="btn btn-neutral float-left" title="Computing the similarity between two BED-files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../interoperability/numpy_operations.html" class="btn btn-neutral float-right" title="Numpy Interoperability" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>