<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Simple quality checking of fastq files &mdash; bionumpy  documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Using Numpy Functionality" href="numpy_operations.html" />
    <link rel="prev" title="FASTQ filtering" href="fastq_filtering.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../readme.html">What is BioNumPy?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation and getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#your-first-bionumpy-program">Your first BioNumPy program</a></li>
</ul>
<p class="caption"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/working_with_big_data.html">Working with big data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Currently supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#implementing-a-new-file-format">Implementing a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html#sort-order">Sort Order</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/rollable_functions.html">Rollable Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/summarization.html">Summarization</a></li>
</ul>
<p class="caption"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../stubs/bionumpy.html">bionumpy</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="bam_handling.html">BAM Handling</a></li>
<li class="toctree-l1"><a class="reference internal" href="fastq_filtering.html">FASTQ filtering</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Simple quality checking of fastq files</a></li>
<li class="toctree-l1"><a class="reference internal" href="numpy_operations.html">Using Numpy Functionality</a></li>
<li class="toctree-l1"><a class="reference internal" href="position_weight_matrix.html">Position Weight Matrix</a></li>
</ul>
<p class="caption"><span class="caption-text">Developer Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/getting_started.html">Getting started and how to contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/setting_up_development_environment.html">Setting up your development environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/testing.html">How we do testing in BioNumPy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/making_examples.html">Making examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/writing_documentation.html">Writing documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/design_principles.html">Design principles and code overview</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Simple quality checking of fastq files</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/fastqc.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="simple-quality-checking-of-fastq-files">
<h1>Simple quality checking of fastq files<a class="headerlink" href="#simple-quality-checking-of-fastq-files" title="Permalink to this headline">¶</a></h1>
<p>In this tutorial we will perform simple quality checking of reads from a fastq file, similarly to what the popular tool FastQC. In addition to BioNumPy, you will also need matplotlib to do some plotting.</p>
<p>We assume you have already followed the introduction part of reading files (see <a class="reference internal" href="../source/reading_files.html#reading-files"><span class="std std-ref">Reading files</span></a>).</p>
<p>We start by importing all we need:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
</pre></div>
</div>
<p>We will be using the <cite>big.fq.gz</cite> file in the example_data folder, but feel free to use any fastq file you like.</p>
<p>The first step is to read our data as chunks:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/big.fq.gz&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
</pre></div>
</div>
<p>Note that we now only have a generator object that will give us chunks when we start iterating over it. No data has been read yet.</p>
<div class="section" id="gc-content">
<h2>GC-content<a class="headerlink" href="#gc-content" title="Permalink to this headline">¶</a></h2>
<p>We want to get the GC-content (as a ratio between 0 and 1) within each sequence in a fastq file, and plot a histogram of these numbers.</p>
<p>For each chunk we read from the file, we get the sequences as a RaggedArray where each row is a sequence. Creating a boolean mask of where we have G or C is then as simple as:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">chunk</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="s2">&quot;G&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">chunk</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="s2">&quot;C&quot;</span><span class="p">)</span> 
</pre></div>
</div>
<p><cite>mask</cite> is now still a ragged array with 1 where we have a C or G and 0 elsewhere.</p>
<p>Getting the GC-content for each read can now be done by taking the mean across the reads (last axis) of this mask:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gc_contents</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span> 
</pre></div>
</div>
<p>If we want to do this across all all sequence chunks, we can create a function that does what we want on one chunk and add the streamable decorator:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nd">@bnp</span><span class="o">.</span><span class="n">streamable</span><span class="p">()</span>
<span class="gp">... </span><span class="k">def</span> <span class="nf">get_gc_content</span><span class="p">(</span><span class="n">reads</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">sequences</span> <span class="o">=</span> <span class="n">reads</span><span class="o">.</span><span class="n">sequence</span>
<span class="gp">... </span>    <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">sequences</span> <span class="o">==</span> <span class="s2">&quot;G&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">sequences</span> <span class="o">==</span> <span class="s2">&quot;C&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>We want to create a histogram of the gc-content values from all chunks. We could call get_gc_content on each chunk, add the results to a list and create a histogram from the final list, but BioNumPy also provides a utility function for creating a histogram from the results from multiple chunks:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">histogram</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">get_gc_content</span><span class="p">(</span><span class="n">reads</span><span class="p">),</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="nb">range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">histogram</span><span class="p">)</span> 
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> 
</pre></div>
</div>
<p>There is some “magic” happening here that might be useful to understand:</p>
<ul class="simple">
<li><p>the &#64;streamable decorator lets us call the get_gc_content on multiple chunks (the result from read_chunks). All it does is to provide an iterable over the results from calling get_gc_content on each chunk.</p></li>
<li><p>the <cite>bnp.histogram</cite> function can take such an iterable and combines the results</p></li>
<li><p>After this code has run, we have iterated over all the chunks in the file, and we need to open the file again and read chunks again if we want to anything else with the file</p></li>
</ul>
</div>
<div class="section" id="histogram-of-base-qualities">
<h2>Histogram of base qualities<a class="headerlink" href="#histogram-of-base-qualities" title="Permalink to this headline">¶</a></h2>
<p>If we want to plot a histogram of all the base qualities in all reads, we can use the builtin <cite>bnp.bincount</cite> function. This function does a numpy bincount on each chunk and combines the results.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nd">@bnp</span><span class="o">.</span><span class="n">streamable</span><span class="p">()</span>
<span class="gp">... </span><span class="k">def</span> <span class="nf">get_base_qualities</span><span class="p">(</span><span class="n">reads</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">reads</span><span class="o">.</span><span class="n">quality</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/big.fq.gz&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">base_quality_bincount</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">get_base_qualities</span><span class="p">(</span><span class="n">reads</span><span class="p">),</span> <span class="n">minlength</span><span class="o">=</span><span class="mi">60</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">base_quality_bincount</span><span class="p">)</span> 
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> 
</pre></div>
</div>
</div>
<div class="section" id="average-base-quality-per-base">
<h2>Average base quality per base<a class="headerlink" href="#average-base-quality-per-base" title="Permalink to this headline">¶</a></h2>
<p>In the GC content histogram example, we saw that we can take the mean the rows (axis=-1). If we instead want to find the average base quality for each position in the reads, we can take the mean across the columns (axis=0). Since the reads may have different lengths, we create a padded matrix filled with zeroes. Note that this means that the average base quality is “wrong” after the minimum read length.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nd">@bnp</span><span class="o">.</span><span class="n">streamable</span><span class="p">()</span>
<span class="gp">... </span><span class="k">def</span> <span class="nf">get_quality_scores_as_matrix</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">limit_at_n_bases</span><span class="o">=</span><span class="mi">150</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">reads</span><span class="o">.</span><span class="n">quality</span><span class="o">.</span><span class="n">as_padded_matrix</span><span class="p">(</span><span class="n">side</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span><span class="p">)[:,</span><span class="mi">0</span><span class="p">:</span><span class="n">limit_at_n_bases</span><span class="p">]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/big.fq.gz&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scores</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">get_quality_scores_as_matrix</span><span class="p">(</span><span class="n">reads</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span> 
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> 
</pre></div>
</div>
<p>Remember to change the limit_at_n_bases depending on your minimum read length (or how much of the reads you want to plot).</p>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="fastq_filtering.html" class="btn btn-neutral float-left" title="FASTQ filtering" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="numpy_operations.html" class="btn btn-neutral float-right" title="Using Numpy Functionality" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>