

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Computing GC content inside genes &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Position Weight Matrix" href="position_weight_matrix.html" />
    <link rel="prev" title="FastQC-like quality-checking of FASTQ files" href="fastqc.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Computing GC content inside genes</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/gc_content.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="computing-gc-content-inside-genes">
<span id="gc-content"></span><h1>Computing GC content inside genes<a class="headerlink" href="#computing-gc-content-inside-genes" title="Link to this heading">ÔÉÅ</a></h1>
<p>This example shows how a classic bioinformatics task like computing GC content inside genes
can be performed intuitively and efficiently using BioNumPy.</p>
<p>It combines two main workhorses of BioNumPy: Sequence arrays for DNA, and Interval arrays for genomic locations.</p>
<p>Sequence arrays are read in from fasta files. Interval arrays are read in from bed files.</p>
<p>The crux of the example is that the gene parts of a DNA sequence are extracted by a single, intuitive indexing operation, simply: <cite>selected_seq = sequence[intervals]</cite>.</p>
<p>The count of a given nulceotide <cite>nn</cite> inside these gene regions is then computed simply by <cite>(selected_seq == nn).sum()</cite>.</p>
<p>Finally, to compute GC content outside genes, one first computes the regions outside genes by inverting a mask of gene regions: <cite>~get_boolean_mask(genes, len(chr1_sequence))</cite></p>
<p>A first version of the example reads in the full sequence and interval data from fasta and bed files that only contain data for a single chromosome: <cite>bnp.open(seq_fn).read()</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Tuple</span>

<span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="kn">from</span> <span class="nn">bionumpy.arithmetics</span> <span class="kn">import</span> <span class="n">get_boolean_mask</span>


<span class="k">def</span> <span class="nf">analyze_within_chromosome</span><span class="p">(</span><span class="n">seq_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">genes_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
    <span class="n">chr1</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">seq_filename</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">genes_filename</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Gene-regions: &quot;</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Fasta: &quot;</span><span class="p">,</span> <span class="n">chr1</span><span class="p">)</span>
    <span class="n">gc_inside</span><span class="p">,</span> <span class="n">gc_outside</span> <span class="o">=</span> <span class="n">gc_inside_and_outside</span><span class="p">(</span><span class="n">chr1</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC inside: </span><span class="si">{</span><span class="n">gc_inside</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">, GC outside: </span><span class="si">{</span><span class="n">gc_outside</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gc_inside</span><span class="p">,</span> <span class="n">gc_outside</span>


<span class="k">def</span> <span class="nf">gc_inside_and_outside</span><span class="p">(</span><span class="n">chr1_sequence</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">EncodedArray</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">Interval</span><span class="p">)</span><span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
    <span class="n">gc_inside</span> <span class="o">=</span> <span class="n">get_gc_content</span><span class="p">(</span><span class="n">chr1_sequence</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span>
    <span class="n">gc_outside</span> <span class="o">=</span> <span class="n">get_gc_content</span><span class="p">(</span><span class="n">chr1_sequence</span><span class="p">,</span> <span class="o">~</span><span class="n">get_boolean_mask</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">chr1_sequence</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">gc_inside</span><span class="p">,</span> <span class="n">gc_outside</span>


<span class="k">def</span> <span class="nf">get_gc_content</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">intervals</span><span class="p">):</span>
    <span class="n">selected_seq</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">intervals</span><span class="p">]</span>
    <span class="n">nn_counts</span> <span class="o">=</span> <span class="p">{</span><span class="n">nn</span><span class="p">:</span> <span class="p">(</span><span class="n">selected_seq</span> <span class="o">==</span> <span class="n">nn</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="s2">&quot;ACTG&quot;</span><span class="p">}</span>
    <span class="n">gc_count</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">nn_counts</span><span class="p">[</span><span class="n">nn</span><span class="p">]</span> <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="s2">&quot;GC&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">gc_count</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">nn_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">analyze_within_chromosome</span><span class="p">(</span><span class="s2">&quot;example_data/gc_test_onechr.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;example_data/gc_bedtest_onechr.bed&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">result</span> <span class="o">==</span> <span class="p">(</span><span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">)</span>

</pre></div>
</div>
<p>A second version still reads in the full data by the <cite>read</cite> function, but adds in how to handle fasta and bed files containing data for multiple chromosomes by splitting the interval data based on the groupby function, and accessing corresponding entries from the read in sequence data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>


<span class="k">def</span> <span class="nf">analyze_across_chromosome</span><span class="p">(</span><span class="n">seq_fn</span><span class="p">,</span> <span class="n">genes_fn</span><span class="p">):</span>
    <span class="n">all_sequence</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">seq_fn</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">all_genes</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">genes_fn</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">genes_by_chr</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">all_genes</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
    <span class="n">nn_counts</span> <span class="o">=</span> <span class="p">{</span><span class="n">nn</span><span class="p">:</span><span class="mi">0</span> <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="s2">&quot;ACTG&quot;</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genes_by_chr</span><span class="p">):</span>
        <span class="n">selected_seq</span> <span class="o">=</span> <span class="n">all_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">genes</span><span class="p">]</span>
        <span class="k">assert</span> <span class="nb">str</span><span class="p">(</span><span class="n">all_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="o">==</span> <span class="nb">chr</span>
        <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="s2">&quot;ACTG&quot;</span><span class="p">:</span>
            <span class="n">nn_counts</span><span class="p">[</span><span class="n">nn</span><span class="p">]</span> <span class="o">+=</span> <span class="p">(</span><span class="n">selected_seq</span> <span class="o">==</span> <span class="n">nn</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">gc_content</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">nn_counts</span><span class="p">[</span><span class="n">nn</span><span class="p">]</span> <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="s2">&quot;GC&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">nn_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC-content inside genes: </span><span class="si">{</span><span class="n">gc_content</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">analyze_across_chromosome</span><span class="p">(</span><span class="s2">&quot;example_data/gc_test_multichr.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;example_data/gc_bedtest_multichr.bed&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>A third version shows a more scalable version of the same code, which uses the <cite>read_chunks</cite> function (see <a class="reference internal" href="../source/reading_files.html#reading-files"><span class="std std-ref">Reading files</span></a>) with the fasta file to get a generator that reads in a single chromosome sequence at a time, allowing it to handle full-genome data with limited memory footprint:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>


<span class="k">def</span> <span class="nf">analyze_across_chromosome</span><span class="p">(</span><span class="n">seq_fn</span><span class="p">,</span> <span class="n">genes_fn</span><span class="p">):</span>
    <span class="n">all_sequence</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">seq_fn</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
    <span class="n">all_genes</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">genes_fn</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">genes_by_chr</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">all_genes</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
    <span class="n">nn_counts</span> <span class="o">=</span> <span class="p">{</span><span class="n">nn</span><span class="p">:</span><span class="mi">0</span> <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="s2">&quot;ACTG&quot;</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">seq</span><span class="p">,</span> <span class="p">(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">all_sequence</span><span class="p">,</span> <span class="n">genes_by_chr</span><span class="p">):</span>
        <span class="k">assert</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="nb">chr</span><span class="p">,</span> <span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">seq</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">chr</span><span class="p">)</span>
        <span class="n">selected_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">genes</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="s2">&quot;ACTG&quot;</span><span class="p">:</span>
            <span class="n">nn_counts</span><span class="p">[</span><span class="n">nn</span><span class="p">]</span> <span class="o">+=</span> <span class="p">(</span><span class="n">selected_seq</span> <span class="o">==</span> <span class="n">nn</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">gc_content</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">nn_counts</span><span class="p">[</span><span class="n">nn</span><span class="p">]</span> <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="s2">&quot;GC&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">nn_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC-content inside genes: </span><span class="si">{</span><span class="n">gc_content</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">analyze_across_chromosome</span><span class="p">(</span><span class="s2">&quot;example_data/gc_test_multichr.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;example_data/gc_bedtest_multichr.bed&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The fourth version shows how to use the <cite>Genome</cite> functionality to conveniently calculate the GC content:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>


<span class="k">def</span> <span class="nf">gc_genomic</span><span class="p">(</span><span class="n">sequence_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">genes_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">Genome</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">sequence_filename</span><span class="p">)</span>
    <span class="n">gene_mask</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_intervals</span><span class="p">(</span><span class="n">genes_filename</span><span class="p">)</span><span class="o">.</span><span class="n">get_mask</span><span class="p">()</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_sequence</span><span class="p">()</span>
    <span class="n">gene_sequences</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">gene_mask</span><span class="p">]</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_encoded</span><span class="p">(</span><span class="n">gene_sequences</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">gc_content</span> <span class="o">=</span> <span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;G&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">counts</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="p">(</span><span class="n">gene_sequences</span><span class="o">.</span><span class="n">size</span> <span class="o">-</span> <span class="n">counts</span><span class="p">[</span><span class="s2">&quot;N&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">gc_content</span>


<span class="k">def</span> <span class="nf">test</span><span class="p">():</span>
    <span class="k">assert</span> <span class="n">gc_genomic</span><span class="p">(</span><span class="s2">&quot;example_data/gc_test_multichr.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;example_data/gc_bedtest_multichr.bed&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="o">/</span> <span class="mi">3</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;GC content </span><span class="si">{</span><span class="n">gc_genomic</span><span class="p">(</span><span class="o">*</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="fastqc.html" class="btn btn-neutral float-left" title="FastQC-like quality-checking of FASTQ files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="position_weight_matrix.html" class="btn btn-neutral float-right" title="Position Weight Matrix" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>