

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Simulating sequence datasets &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Computing the similarity between two BED-files" href="similarity_measures.html" />
    <link rel="prev" title="Position Weight Matrix" href="position_weight_matrix.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Simulating sequence datasets</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/set_of_sequences.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="simulating-sequence-datasets">
<h1>Simulating sequence datasets<a class="headerlink" href="#simulating-sequence-datasets" title="Link to this heading">ÔÉÅ</a></h1>
<p>Simulating sequences is very straightforward in BioNumPy. Since a sequence arrays have underlying numeric representations that are easy to relate to, one can directly simulate such an underlying representation using any numeric simulation procedure. In addition, bionumpy provides convenience functions that allows to directly simulate sequence data without having to think about any underlying numeric representation. This is the focus of the current tutorial. The function simulate_sequences is an easy way to simulate a set of sequences, by simply specifying a desired alphabet and a dictionary with desired sequence ids as keys and the desired length of each such sequence as value (here simulating 20 sequences with length 10..30):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.simulate</span> <span class="kn">import</span> <span class="n">simulate_sequences</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.sequence</span> <span class="kn">import</span> <span class="n">match_string</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">named_seqs</span> <span class="o">=</span> <span class="n">simulate_sequences</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="p">{</span><span class="sa">f</span><span class="s1">&#39;s</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span><span class="mi">10</span><span class="o">+</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">20</span><span class="p">)},</span> <span class="n">rng</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">named_seqs</span>
<span class="go">SequenceEntry with 20 entries</span>
<span class="go">                     name                 sequence</span>
<span class="go">                       s0               CGTTAATTAC</span>
<span class="go">                       s1              TCCTCCGGAAT</span>
<span class="go">                       s2             TTGTCCTACACT</span>
<span class="go">                       s3            ACCTAGCATACCC</span>
<span class="go">                       s4           ATGTAGCGTCGACT</span>
<span class="go">                       s5          CGCACGCTCGTTCAG</span>
<span class="go">                       s6         GTCCACGTTAGTCCTG</span>
<span class="go">                       s7        GGGTTAAGTAGTTTAGT</span>
<span class="go">                       s8       CACAATGTTTCCGCTATG</span>
<span class="go">                       s9      CGCTTCCAGGTTTTTAACC</span>
</pre></div>
</div>
<p>One can now easily do a variety of analyses on these simulated sequences, e.g. compute the GC content per simulated sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="n">named_seqs</span><span class="o">.</span><span class="n">sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_content_per_seq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">((</span><span class="n">seqs</span><span class="o">==</span><span class="s1">&#39;C&#39;</span><span class="p">)</span><span class="o">|</span><span class="p">(</span><span class="n">seqs</span><span class="o">==</span><span class="s1">&#39;G&#39;</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_content_per_seq</span>
<span class="go">array([0.3       , 0.54545455, 0.41666667, 0.53846154, 0.5       ,</span>
<span class="go">       0.66666667, 0.5625    , 0.35294118, 0.44444444, 0.47368421,</span>
<span class="go">       0.5       , 0.38095238, 0.68181818, 0.30434783, 0.66666667,</span>
<span class="go">       0.52      , 0.46153846, 0.44444444, 0.53571429, 0.51724138])</span>
</pre></div>
</div>
<p>If desired, such computed values per sequence can easily be added back as an additional column of the bionumpy data structure:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">named_seqs</span> <span class="o">=</span> <span class="n">named_seqs</span><span class="o">.</span><span class="n">add_fields</span><span class="p">({</span><span class="s1">&#39;gc&#39;</span><span class="p">:</span><span class="n">gc_content_per_seq</span><span class="p">},</span> <span class="p">{</span><span class="s1">&#39;gc&#39;</span><span class="p">:</span><span class="nb">float</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">named_seqs</span>
<span class="go">DynamicSequenceEntry with 20 entries</span>
<span class="go">                     name                 sequence                       gc</span>
<span class="go">                       s0               CGTTAATTAC                      0.3</span>
<span class="go">                       s1              TCCTCCGGAAT       0.5454545454545454</span>
<span class="go">                       s2             TTGTCCTACACT       0.4166666666666667</span>
<span class="go">                       s3            ACCTAGCATACCC       0.5384615384615384</span>
<span class="go">                       s4           ATGTAGCGTCGACT                      0.5</span>
<span class="go">                       s5          CGCACGCTCGTTCAG       0.6666666666666666</span>
<span class="go">                       s6         GTCCACGTTAGTCCTG                   0.5625</span>
<span class="go">                       s7        GGGTTAAGTAGTTTAGT      0.35294117647058826</span>
<span class="go">                       s8       CACAATGTTTCCGCTATG       0.4444444444444444</span>
<span class="go">                       s9      CGCTTCCAGGTTTTTAACC      0.47368421052631576</span>
</pre></div>
</div>
<dl>
<dt>We can also easily apply a variety of built-in bionumpy functionality on our simulated sequences:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ac_hits</span> <span class="o">=</span> <span class="n">match_string</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="s2">&quot;AC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ac_hit_sums</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">ac_hits</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ac_hit_sums</span>
<span class="go">array([1, 0, 2, 2, 1, 1, 1, 0, 1, 1, 1, 1, 2, 1, 2, 2, 2, 0, 1, 1])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">named_seqs</span> <span class="o">=</span> <span class="n">named_seqs</span><span class="o">.</span><span class="n">add_fields</span><span class="p">({</span><span class="s1">&#39;ac_hits&#39;</span><span class="p">:</span><span class="n">ac_hit_sums</span><span class="p">},</span> <span class="p">{</span><span class="s1">&#39;ac_hits&#39;</span><span class="p">:</span><span class="nb">int</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">named_seqs</span>
<span class="go">DynamicSequenceEntry with 20 entries</span>
<span class="go">                     name                 sequence                       gc                  ac_hits</span>
<span class="go">                       s0               CGTTAATTAC                      0.3                        1</span>
<span class="go">                       s1              TCCTCCGGAAT       0.5454545454545454                        0</span>
<span class="go">                       s2             TTGTCCTACACT       0.4166666666666667                        2</span>
<span class="go">                       s3            ACCTAGCATACCC       0.5384615384615384                        2</span>
<span class="go">                       s4           ATGTAGCGTCGACT                      0.5                        1</span>
<span class="go">                       s5          CGCACGCTCGTTCAG       0.6666666666666666                        1</span>
<span class="go">                       s6         GTCCACGTTAGTCCTG                   0.5625                        1</span>
<span class="go">                       s7        GGGTTAAGTAGTTTAGT      0.35294117647058826                        0</span>
<span class="go">                       s8       CACAATGTTTCCGCTATG       0.4444444444444444                        1</span>
<span class="go">                       s9      CGCTTCCAGGTTTTTAACC      0.47368421052631576                        1</span>
</pre></div>
</div>
</dd>
</dl>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="position_weight_matrix.html" class="btn btn-neutral float-left" title="Position Weight Matrix" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="similarity_measures.html" class="btn btn-neutral float-right" title="Computing the similarity between two BED-files" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>