

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bionumpy.genomic_data.genomic_track &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../menu.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bionumpy.genomic_data.genomic_track</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bionumpy.genomic_data.genomic_track</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">abc</span> <span class="kn">import</span> <span class="n">abstractmethod</span>
<span class="kn">from</span> <span class="nn">.genomic_data</span> <span class="kn">import</span> <span class="n">GenomicData</span>
<span class="kn">from</span> <span class="nn">.genome_context_base</span> <span class="kn">import</span> <span class="n">GenomeContextBase</span>
<span class="kn">from</span> <span class="nn">..computation_graph</span> <span class="kn">import</span> <span class="n">StreamNode</span><span class="p">,</span> <span class="n">ComputationNode</span>
<span class="kn">from</span> <span class="nn">..datatypes</span> <span class="kn">import</span> <span class="n">Interval</span><span class="p">,</span> <span class="n">BedGraph</span>
<span class="kn">from</span> <span class="nn">..arithmetics.intervals</span> <span class="kn">import</span> <span class="n">GenomicRunLengthArray</span>
<span class="kn">from</span> <span class="nn">npstructures</span> <span class="kn">import</span> <span class="n">RunLengthRaggedArray</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Callable</span>


<div class="viewcode-block" id="GenomicArray">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.GenomicArray">[docs]</a>
<span class="k">class</span> <span class="nc">GenomicArray</span><span class="p">(</span><span class="n">GenomicData</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Class for representing data on a genome.</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genome_context</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomeContextBase</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span>

    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">__array_ufunc__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ufunc</span><span class="p">:</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">*</span><span class="n">inputs</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicArray&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">NotImplemented</span>

    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">__array_function__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">func</span><span class="p">:</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">types</span><span class="p">:</span> <span class="n">List</span><span class="p">,</span> <span class="n">args</span><span class="p">:</span> <span class="n">List</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Any</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">NotImplemented</span>

    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">sum</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">axis</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">NotImplemented</span>

    <span class="k">def</span> <span class="nf">to_bedgraph</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BedGraph</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">NotImplemented</span>

<div class="viewcode-block" id="GenomicArray.from_global_data">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.GenomicArray.from_global_data">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_global_data</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">global_pileup</span><span class="p">:</span> <span class="n">GenomicRunLengthArray</span><span class="p">,</span>
                         <span class="n">genome_context</span><span class="p">:</span> <span class="n">GenomeContextBase</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicArray&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create the genomic array from data represented on a flat/concatenated genome.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        global_pileup : GenomicRunLengthArray</span>
<span class="sd">            Array over the concatenated genome</span>
<span class="sd">        genome_context : GenomeContextBase</span>
<span class="sd">            The genome context</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        &#39;GenomicArray&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">GenomicArrayGlobal</span><span class="p">(</span><span class="n">global_pileup</span><span class="p">,</span> <span class="n">genome_context</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicArray.from_bedgraph">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.GenomicArray.from_bedgraph">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_bedgraph</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">bedgraph</span><span class="p">:</span> <span class="n">BedGraph</span><span class="p">,</span> <span class="n">genome_context</span><span class="p">:</span> <span class="n">GenomeContextBase</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicData&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create a genomic array from a bedgraph and genomic context</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        bedgraph : BedGraph</span>
<span class="sd">            Bedgraph with data along the genome</span>
<span class="sd">        genome_context : GenomeContextBase</span>
<span class="sd">            The genome context</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        &#39;GenomicData&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">bedgraph</span><span class="p">,</span> <span class="n">BedGraph</span><span class="p">):</span>
            <span class="n">go</span> <span class="o">=</span> <span class="n">genome_context</span><span class="o">.</span><span class="n">global_offset</span>
            <span class="n">gi</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">from_local_interval</span><span class="p">(</span><span class="n">bedgraph</span><span class="p">)</span>
            <span class="n">rle</span> <span class="o">=</span> <span class="n">GenomicRunLengthArray</span><span class="o">.</span><span class="n">from_bedgraph</span><span class="p">(</span><span class="n">gi</span><span class="p">,</span> <span class="n">go</span><span class="o">.</span><span class="n">total_size</span><span class="p">())</span>
            <span class="k">return</span> <span class="bp">cls</span><span class="o">.</span><span class="n">from_global_data</span><span class="p">(</span><span class="n">rle</span><span class="p">,</span> <span class="n">genome_context</span><span class="p">)</span>
        <span class="n">filled</span> <span class="o">=</span> <span class="n">genome_context</span><span class="o">.</span><span class="n">iter_chromosomes</span><span class="p">(</span><span class="n">bedgraph</span><span class="p">,</span> <span class="n">BedGraph</span><span class="p">)</span>
        <span class="c1"># filled = fill_grouped(groupby(bedgraph, &#39;chromosome&#39;), chrom_sizes.keys(), BedGraph)</span>
        <span class="n">interval_stream</span> <span class="o">=</span> <span class="n">StreamNode</span><span class="p">(</span><span class="n">filled</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">GenomicArrayNode</span><span class="p">(</span><span class="n">ComputationNode</span><span class="p">(</span><span class="n">GenomicRunLengthArray</span><span class="o">.</span><span class="n">from_bedgraph</span><span class="p">,</span>
                                                <span class="p">[</span><span class="n">interval_stream</span><span class="p">,</span>
                                                 <span class="n">StreamNode</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">genome_context</span><span class="o">.</span><span class="n">chrom_sizes</span><span class="o">.</span><span class="n">values</span><span class="p">()))]),</span>
                                <span class="n">genome_context</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">_get_intervals_from_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">data</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="nb">bool</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">Interval</span><span class="p">([</span><span class="n">name</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">starts</span><span class="p">),</span>
                            <span class="n">data</span><span class="o">.</span><span class="n">starts</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">ends</span><span class="p">)[</span><span class="n">data</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">BedGraph</span><span class="p">([</span><span class="n">name</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">starts</span><span class="p">),</span>
                            <span class="n">data</span><span class="o">.</span><span class="n">starts</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">ends</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">values</span><span class="p">)</span></div>



<span class="k">class</span> <span class="nc">GenomicArrayGlobal</span><span class="p">(</span><span class="n">GenomicArray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">lib</span><span class="o">.</span><span class="n">mixins</span><span class="o">.</span><span class="n">NDArrayOperatorsMixin</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Class for holding the the enitre genomic array in memory</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">global_track</span><span class="p">:</span> <span class="n">GenomicRunLengthArray</span><span class="p">,</span> <span class="n">genome_context</span><span class="p">:</span> <span class="n">GenomeContextBase</span><span class="p">):</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">global_track</span><span class="p">,</span> <span class="n">GenomicRunLengthArray</span><span class="p">),</span> <span class="n">global_track</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span> <span class="o">=</span> <span class="n">global_track</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span> <span class="o">=</span> <span class="n">genome_context</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">dtype</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">dtype</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="o">.</span><span class="n">dtype</span>

    <span class="k">def</span> <span class="nf">_index_boolean</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idx</span><span class="p">):</span>
        <span class="k">assert</span> <span class="n">idx</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="nb">bool</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">GenomicArrayGlobal</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="p">[</span><span class="n">idx</span><span class="o">.</span><span class="n">_global_track</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">sum</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;Sum the data in the array&#39;&#39;&#39;</span>
        <span class="k">assert</span> <span class="n">axis</span> <span class="ow">is</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">extract_chromsome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span>
        <span class="n">GenomicRunLengthArray</span><span class="p">,</span> <span class="n">RunLengthRaggedArray</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the data on one or more chromosomes</span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chromosome : Union[str, List[str]]</span>
<span class="sd">            A chromosome name or a list of chromosome names</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Union[GenomicRunLengthArray, RunLengthRaggedArray]</span>
<span class="sd">            A runlengtharray for a single chromosome or a RunLengthRaggedArray for a list of chromosomes</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
        <span class="n">offset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">global_offset</span><span class="o">.</span><span class="n">get_offset</span><span class="p">([</span><span class="n">chromosome</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="p">[</span><span class="n">offset</span><span class="p">:</span><span class="n">offset</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">global_offset</span><span class="o">.</span><span class="n">get_size</span><span class="p">([</span><span class="n">chromosome</span><span class="p">])[</span><span class="mi">0</span><span class="p">]]</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">chrom_sizes</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)):</span>
            <span class="n">lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="bp">self</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">chrom_sizes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
            <span class="n">lines</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="s1">&#39;...&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">to_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">GenomicRunLengthArray</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Convert the genomic array to a dict of arrays with chromosomes as keys</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Dict[str, GenomicRunLengthArray]</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">go</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">global_offset</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">names</span><span class="p">()</span>
        <span class="n">offsets</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">get_offset</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
        <span class="n">sizes</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">get_size</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="p">[</span><span class="n">offset</span><span class="p">:</span><span class="n">offset</span> <span class="o">+</span> <span class="n">size</span><span class="p">]</span><span class="o">.</span><span class="n">to_array</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">offset</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="n">offsets</span><span class="p">,</span> <span class="n">sizes</span><span class="p">)}</span>

    <span class="k">def</span> <span class="nf">__array_ufunc__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ufunc</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ufunc</span><span class="p">,</span> <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">*</span><span class="n">inputs</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Handle numpy ufuncs on the genomic array</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ufunc: np.ufunc</span>
<span class="sd">        method: str</span>
<span class="sd">            How to call the ufunc</span>
<span class="sd">        inputs: List</span>
<span class="sd">            Args for  the ufunc</span>
<span class="sd">        kwargs: Dict</span>
<span class="sd">            Additional arguments</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicArrayGlobal</span>

<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">inputs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span><span class="o">.</span><span class="n">_global_track</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">GenomicArrayGlobal</span><span class="p">)</span> <span class="k">else</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">inputs</span><span class="p">]</span>
        <span class="n">r</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="o">.</span><span class="n">__array_ufunc__</span><span class="p">(</span><span class="n">ufunc</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="o">*</span><span class="n">inputs</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__array_function__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">func</span><span class="p">:</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">types</span><span class="p">:</span> <span class="n">List</span><span class="p">,</span> <span class="n">args</span><span class="p">:</span> <span class="n">List</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Handles any numpy array functions called on a raggedarray</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        func : callable</span>
<span class="sd">        types : List</span>
<span class="sd">        args : List</span>
<span class="sd">        kwargs : Dict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">args</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span><span class="o">.</span><span class="n">_global_track</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">GenomicArrayGlobal</span><span class="p">)</span> <span class="k">else</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">args</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">func</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">func</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">NotImplemented</span>

    <span class="k">def</span> <span class="nf">get_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">Interval</span><span class="p">,</span> <span class="n">BedGraph</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;The data of the array represented in BNPDataClass</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Union[Interval, BedGraph]</span>
<span class="sd">            If boolean mask,return the intervals where True. Else a bedgraph containing the data</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">go</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">global_offset</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">names</span><span class="p">()</span>
        <span class="n">starts</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">get_offset</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
        <span class="n">stops</span> <span class="o">=</span> <span class="n">starts</span> <span class="o">+</span> <span class="n">go</span><span class="o">.</span><span class="n">get_size</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
        <span class="n">intervals_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="n">starts</span><span class="p">,</span> <span class="n">stops</span><span class="p">):</span>
            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="p">,</span> <span class="n">GenomicRunLengthArray</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">]</span>
            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">GenomicRunLengthArray</span><span class="p">)</span>
            <span class="n">intervals_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_intervals_from_data</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">data</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">intervals_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">extract_intervals</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">intervals</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Interval</span><span class="p">,</span> <span class="s1">&#39;GenomicIntervals&#39;</span><span class="p">],</span>
                          <span class="n">stranded</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RunLengthRaggedArray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract the data contained in a set of intervals</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        intervals : Interval</span>
<span class="sd">        stranded : bool</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        RunLengthRaggedArray</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">global_intervals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">global_offset</span><span class="o">.</span><span class="n">from_local_interval</span><span class="p">(</span><span class="n">intervals</span><span class="p">)</span>
        <span class="n">rle</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="p">[</span><span class="n">global_intervals</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">stranded</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">rle</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">rle</span><span class="p">[:,</span> <span class="p">::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">((</span><span class="n">intervals</span><span class="o">.</span><span class="n">strand</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">)[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">],</span>
                        <span class="n">rle</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">extract_locations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">locations</span><span class="p">:</span> <span class="s1">&#39;GenomicLocation&#39;</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RunLengthRaggedArray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract the data contained in a set of intervals</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        intervals : Interval</span>
<span class="sd">        stranded : bool</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        RunLengthRaggedArray</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="n">stranded</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;Stranded not implemented for locations&#39;</span>
        <span class="n">global_intervals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">global_offset</span><span class="o">.</span><span class="n">from_local_coordinates</span><span class="p">(</span><span class="n">locations</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span>
                                                                                     <span class="n">locations</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_global_track</span><span class="p">[</span><span class="n">global_intervals</span><span class="p">]</span>
        <span class="c1"># if not stranded:</span>
        <span class="c1">#     return rle</span>
        <span class="c1"># r = rle[:, ::-1]</span>
        <span class="c1"># return np.where((intervals.strand.ravel() == &#39;+&#39;)[:, np.newaxis],</span>
        <span class="c1">#                 rle, r)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_dict</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">GenomicRunLengthArray</span><span class="p">],</span> <span class="n">genome_context</span><span class="p">:</span> <span class="n">GenomeContextBase</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicData&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create genomic data from a dict of data with chromosomes as keys</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        d : Dict[str, GenomicRunLengthArray]</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        &#39;GenomicData&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># chrom_sizes = {name: array.size for name, array in d.items()}</span>
        <span class="n">array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="n">genome_context</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_stream</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">stream</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">GenomicRunLengthArray</span><span class="p">]],</span>
                    <span class="n">genome_context</span><span class="p">:</span> <span class="n">GenomeContextBase</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicData&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Create a genomic array from a stream of data</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        stream: Iterable[Tuple[str, GenomicRunLengthArray]]</span>
<span class="sd">        genome_context: GenomeContextBase</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicData</span>

<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">stream</span><span class="p">))</span>


<span class="k">class</span> <span class="nc">GenomicArrayNode</span><span class="p">(</span><span class="n">GenomicArray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">lib</span><span class="o">.</span><span class="n">mixins</span><span class="o">.</span><span class="n">NDArrayOperatorsMixin</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Class for representing a genomic array, but only keeping data for one </span>
<span class="sd">    chromosome in memory at a time</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;GTN:&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_run_length_node</span><span class="p">)</span>

    <span class="fm">__repr__</span> <span class="o">=</span> <span class="fm">__str__</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">run_length_node</span><span class="p">:</span> <span class="n">ComputationNode</span><span class="p">,</span> <span class="n">genome_context</span><span class="p">:</span> <span class="n">GenomeContextBase</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_run_length_node</span> <span class="o">=</span> <span class="n">run_length_node</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_chrom_name_node</span> <span class="o">=</span> <span class="n">StreamNode</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">genome_context</span><span class="o">.</span><span class="n">chrom_sizes</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span> <span class="o">=</span> <span class="n">genome_context</span>

    <span class="k">def</span> <span class="nf">__array_ufunc__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ufunc</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ufunc</span><span class="p">,</span> <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">*</span><span class="n">inputs</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">args</span> <span class="o">=</span> <span class="p">[</span><span class="n">gtn</span><span class="o">.</span><span class="n">_run_length_node</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gtn</span><span class="p">,</span> <span class="n">GenomicArrayNode</span><span class="p">)</span> <span class="k">else</span> <span class="n">gtn</span> <span class="k">for</span> <span class="n">gtn</span> <span class="ow">in</span> <span class="n">inputs</span><span class="p">]</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="n">ufunc</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ComputationNode</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the data from the array in a lazy BNPDataClass</span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        ComputationNode</span>
<span class="sd">            A lazy compution node representing the data from the array</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">ComputationNode</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_intervals_from_data</span><span class="p">,</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_chrom_name_node</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_run_length_node</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">_index_boolean</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idx</span><span class="p">):</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">GenomicArrayNode</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ComputationNode</span><span class="p">(</span><span class="k">lambda</span> <span class="n">a</span><span class="p">,</span> <span class="n">i</span><span class="p">:</span> <span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_run_length_node</span><span class="p">,</span> <span class="n">idx</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">extract_intervals</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">intervals</span><span class="p">:</span> <span class="n">Interval</span><span class="p">,</span> <span class="n">stranded</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ComputationNode</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract the data on a set of intervals.</span>

<span class="sd">        Returns a `ComputationNode` representing the extracted data.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        intervals : Interval</span>
<span class="sd">            The set of intervals to extract data from</span>
<span class="sd">        stranded : bool</span>
<span class="sd">            Whether or not to treat the intervals as stranded</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        ComputationNode:</span>
<span class="sd">            A computation node representing the raggedarray containing data from all the intervals</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span> <span class="nf">stranded_func</span><span class="p">(</span><span class="n">ra</span><span class="p">:</span> <span class="n">GenomicRunLengthArray</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">stop</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">strand</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>

            <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">stop</span> <span class="o">&lt;=</span> <span class="n">ra</span><span class="o">.</span><span class="n">size</span><span class="p">),</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">stop</span><span class="p">),</span> <span class="n">ra</span><span class="o">.</span><span class="n">size</span><span class="p">)</span>
            <span class="n">rle</span> <span class="o">=</span> <span class="n">ra</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">]</span>
            <span class="n">r</span> <span class="o">=</span> <span class="n">rle</span><span class="p">[:,</span> <span class="p">::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">((</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">)[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">],</span>
                            <span class="n">rle</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">unstranded_func</span><span class="p">(</span><span class="n">ra</span><span class="p">:</span> <span class="n">GenomicRunLengthArray</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">stop</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">stop</span> <span class="o">&lt;=</span> <span class="n">ra</span><span class="o">.</span><span class="n">size</span><span class="p">),</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">stop</span><span class="p">),</span> <span class="n">ra</span><span class="o">.</span><span class="n">size</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">ra</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">]</span>

        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_context</span><span class="o">.</span><span class="n">is_compatible</span><span class="p">(</span><span class="n">intervals</span><span class="o">.</span><span class="n">genome_context</span><span class="p">),</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_context</span><span class="p">,</span> <span class="n">intervals</span><span class="o">.</span><span class="n">genome_context</span><span class="p">)</span>
        <span class="n">intervals</span> <span class="o">=</span> <span class="n">intervals</span><span class="o">.</span><span class="n">as_stream</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">stranded</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ComputationNode</span><span class="p">(</span><span class="n">stranded_func</span><span class="p">,</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_run_length_node</span><span class="p">,</span>
                                                   <span class="n">intervals</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                                                   <span class="n">intervals</span><span class="o">.</span><span class="n">stop</span><span class="p">,</span>
                                                   <span class="n">intervals</span><span class="o">.</span><span class="n">strand</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">ComputationNode</span><span class="p">(</span><span class="n">unstranded_func</span><span class="p">,</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_run_length_node</span><span class="p">,</span> <span class="n">intervals</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">intervals</span><span class="o">.</span><span class="n">stop</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">extract_chromsome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicData&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Extract the data for a chromosome</span>
<span class="sd">        Unimplemented</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">assert</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">from_stream</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">stream</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">GenomicRunLengthArray</span><span class="p">]],</span>
                    <span class="n">genome_context</span><span class="p">:</span> <span class="n">GenomeContextBase</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicData&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Create a genomic array from a stream of data</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        stream: Iterable[Tuple[str, GenomicRunLengthArray]]</span>
<span class="sd">        genome_context: GenomeContextBase</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicData</span>

<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">stream_node</span> <span class="o">=</span> <span class="n">StreamNode</span><span class="p">((</span><span class="n">array</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">array</span> <span class="ow">in</span> <span class="n">stream</span><span class="p">))</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">stream_node</span><span class="p">,</span> <span class="n">genome_context</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_dict</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">GenomicRunLengthArray</span><span class="p">],</span> <span class="n">genome_context</span><span class="p">:</span> <span class="n">GenomeContextBase</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;GenomicData&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create genomic data from a dict of data with chromosomes as keys</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        d : Dict[str, GenomicRunLengthArray]</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        &#39;GenomicData&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># chrom_sizes = {name: array.size for name, array in d.items()}</span>
        <span class="n">stream_node</span> <span class="o">=</span> <span class="n">StreamNode</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">stream_node</span><span class="p">,</span> <span class="n">genome_context</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">to_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;Unimplemented&#39;&#39;&#39;</span>
        <span class="k">assert</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">sum</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;Sum the data in the array&#39;&#39;&#39;</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_run_length_node</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__array_function__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">func</span><span class="p">:</span> <span class="nb">callable</span><span class="p">,</span> <span class="n">types</span><span class="p">:</span> <span class="n">List</span><span class="p">,</span> <span class="n">args</span><span class="p">:</span> <span class="n">List</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Handles any numpy array functions called on a raggedarray</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        func : callable</span>
<span class="sd">        types : List</span>
<span class="sd">        args : List</span>
<span class="sd">        kwargs : Dict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">func</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_run_length_node</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">NotImplemented</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>