

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bionumpy.genomic_data.genome &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../menu.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bionumpy.genomic_data.genome</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bionumpy.genomic_data.genome</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">PurePath</span>
<span class="kn">from</span> <span class="nn">..io</span> <span class="kn">import</span> <span class="n">bnp_open</span><span class="p">,</span> <span class="n">Bed6Buffer</span><span class="p">,</span> <span class="n">BedBuffer</span><span class="p">,</span> <span class="n">open_indexed</span>
<span class="kn">from</span> <span class="nn">..io.files</span> <span class="kn">import</span> <span class="n">buffer_types</span><span class="p">,</span> <span class="n">BamBuffer</span><span class="p">,</span> <span class="n">BamIntervalBuffer</span>
<span class="kn">from</span> <span class="nn">..io.indexed_files</span> <span class="kn">import</span> <span class="n">IndexBuffer</span><span class="p">,</span> <span class="n">create_index</span>
<span class="kn">from</span> <span class="nn">.genomic_track</span> <span class="kn">import</span> <span class="n">GenomicArray</span>
<span class="kn">from</span> <span class="nn">.genomic_intervals</span> <span class="kn">import</span> <span class="n">GenomicIntervals</span><span class="p">,</span> <span class="n">GenomicLocation</span><span class="p">,</span> <span class="n">LocationEntry</span>
<span class="kn">from</span> <span class="nn">.genomic_sequence</span> <span class="kn">import</span> <span class="n">GenomicSequence</span>
<span class="kn">from</span> <span class="nn">.annotation</span> <span class="kn">import</span> <span class="n">GenomicAnnotation</span>
<span class="kn">from</span> <span class="nn">.genome_context</span> <span class="kn">import</span> <span class="n">GenomeContext</span><span class="p">,</span> <span class="n">keep_all</span><span class="p">,</span> <span class="n">ignore_underscores</span>
<span class="kn">from</span> <span class="nn">..encoded_array</span> <span class="kn">import</span> <span class="n">as_encoded_array</span>
<span class="kn">from</span> <span class="nn">..bnpdataclass</span> <span class="kn">import</span> <span class="n">replace</span><span class="p">,</span> <span class="n">BNPDataClass</span>
<span class="kn">from</span> <span class="nn">..datatypes</span> <span class="kn">import</span> <span class="n">BedGraph</span><span class="p">,</span> <span class="n">Interval</span>
<span class="kn">from</span> <span class="nn">..util.formating</span> <span class="kn">import</span> <span class="n">table</span>


<div class="viewcode-block" id="Genome">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome">[docs]</a>
<span class="k">class</span> <span class="nc">Genome</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    The entry point for working with genomic data. Supports reading data directly from</span>
<span class="sd">    file and into GenomicData objects, or creating GenomicData objects from `BNPDataClass` objects, </span>
<span class="sd">    such `Interval`, `BedGraph` or `LocationEntry`.</span>

<span class="sd">    Supports both in memory or streamed/lazy data, controlled with the `stream` keyword. For streamed</span>
<span class="sd">    data it requires that the data in the file uses the same chromosome ordering as the genome. This</span>
<span class="sd">    will usually be the same as either what is provided in the `chrom.sizes` file, or a simple alphabetic</span>
<span class="sd">    ordering. If sort order errors occur, try using `sort_names=True` when creating the `Genome` object.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom_sizes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="n">fasta_filename</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">sort_names</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">filter_function</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="kc">True</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="p">,</span> <span class="n">GenomeContext</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span> <span class="o">=</span> <span class="n">chrom_sizes</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">sort_names</span><span class="p">:</span>
                <span class="n">chrom_sizes</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">chrom_sizes</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="o">.</span><span class="n">keys</span><span class="p">())}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span> <span class="o">=</span> <span class="n">GenomeContext</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="p">,</span> <span class="n">filter_function</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_fasta_filename</span> <span class="o">=</span> <span class="n">fasta_filename</span>

<div class="viewcode-block" id="Genome.with_ignored_added">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.with_ignored_added">[docs]</a>
    <span class="k">def</span> <span class="nf">with_ignored_added</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignored</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="s1">&#39;Genome&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Make a new GenomeContext with additional ignored chromosomes. This is useful for allowing but ignoring</span>
<span class="sd">        chromosome names that are not in the origin genome.</span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ignored: Iterable[str]</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Genome</span>

<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">with_ignored_added</span><span class="p">(</span><span class="n">ignored</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fasta_filename</span><span class="p">)</span></div>


<div class="viewcode-block" id="Genome.from_dict">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.from_dict">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_dict</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">chrom_sizes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;Genome&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Create a Genome object from a dictionary of chromosome sizes</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom_sizes: Dict[str, int]</span>
<span class="sd">        args: Additional args to be passed to the Genome constructor</span>
<span class="sd">        kwargs: Additional kwargs to be passed to the Genome constructor</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Genome</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; bnp.Genome.from_dict({&#39;chr1&#39;: 1000, &#39;chr2&#39;: 2000})</span>
<span class="sd">        Genome([&#39;chr1&#39;, &#39;chr2&#39;])</span>

<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>


<div class="viewcode-block" id="Genome.from_file">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.from_file">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_file</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sort_names</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">filter_function</span><span class="o">=</span><span class="n">ignore_underscores</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;Genome&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Read genome information from a &#39;chrom.sizes&#39; or &#39;fa.fai&#39; file</span>

<span class="sd">        File should contain rows of names and lengths of chromosomes.</span>
<span class="sd">        If a fasta file is provided, a fasta index will be written to &#39;fa.fai&#39;</span>
<span class="sd">        if not already theree. The sequence will then be available through, &#39;genome.read_sequence()&#39;</span>
<span class="sd">        without any &#39;filename&#39; parameter.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cls :</span>
<span class="sd">        filename : str</span>
<span class="sd">        sort_names : bool Whether or not to sort the chromosome names</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        &#39;Genome&#39;</span>
<span class="sd">            A `Genome` object created from the read chromosome sizes</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; bnp.Genome.from_file(&#39;example_data/hg38.chrom.sizes&#39;)</span>
<span class="sd">        Genome([&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;, &#39;chr4&#39;, &#39;chr5&#39;, &#39;chr6&#39;, &#39;chr7&#39;, &#39;chr8&#39;, &#39;chr9&#39;, &#39;chr10&#39;, &#39;...&#39;])</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">path</span> <span class="o">=</span> <span class="n">PurePath</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="n">suffix</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">suffixes</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">index_file_name</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="n">path</span><span class="o">.</span><span class="n">suffix</span> <span class="o">+</span> <span class="s2">&quot;.fai&quot;</span><span class="p">)</span>
        <span class="n">fasta_filename</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">suffix</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;.fasta&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">index_file_name</span><span class="p">):</span>
                <span class="n">bnp_open</span><span class="p">(</span><span class="n">index_file_name</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">IndexBuffer</span><span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">create_index</span><span class="p">(</span><span class="n">path</span><span class="p">))</span>
            <span class="n">fasta_filename</span> <span class="o">=</span> <span class="n">filename</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">index_file_name</span>

        <span class="n">split_lines</span> <span class="o">=</span> <span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[:</span><span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">({</span><span class="n">name</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">length</span><span class="p">)</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">length</span> <span class="ow">in</span> <span class="n">split_lines</span><span class="p">},</span> <span class="n">fasta_filename</span><span class="o">=</span><span class="n">fasta_filename</span><span class="p">,</span> <span class="n">sort_names</span><span class="o">=</span><span class="n">sort_names</span><span class="p">,</span> <span class="n">filter_function</span><span class="o">=</span><span class="n">filter_function</span><span class="p">)</span></div>


    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">f</span> <span class="o">=</span> <span class="n">bnp_open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">buffer_type</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">stream</span><span class="p">:</span>
            <span class="n">content</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">content</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">content</span>

<div class="viewcode-block" id="Genome.get_track">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.get_track">[docs]</a>
    <span class="k">def</span> <span class="nf">get_track</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bedgraph</span><span class="p">:</span> <span class="n">BedGraph</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomicArray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create a `GenomicArray` for the data in the bedgraph</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        bedgraph : BedGraph</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicArray</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; bedgraph = bnp.datatypes.BedGraph(chromosome=[&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr2&#39;], start=[0, 10, 0], stop=[5, 15, 5], value=[1, 2, 3])</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_dict({&#39;chr1&#39;: 20, &#39;chr2&#39;: 10})</span>
<span class="sd">        &gt;&gt;&gt; genome.get_track(bedgraph)</span>
<span class="sd">        chr1: [1 1 1 1 1 0 0 0 0 0 2 2 2 2 2 0 0 0 0 0]</span>
<span class="sd">        chr2: [3 3 3 3 3 0 0 0 0 0]</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bedgraph</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mask_data_on_extra_chromosomes</span><span class="p">(</span><span class="n">bedgraph</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">GenomicArray</span><span class="o">.</span><span class="n">from_bedgraph</span><span class="p">(</span><span class="n">bedgraph</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="p">)</span></div>


<div class="viewcode-block" id="Genome.read_track">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.read_track">[docs]</a>
    <span class="k">def</span> <span class="nf">read_track</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">stream</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomicArray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Read a bedgraph from file and convert it to a `GenomicArray`</span>

<span class="sd">        If `stream` is `True` then read the bedgraph in chunks and get</span>
<span class="sd">        a lazy evaluated GenomicArray</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            Filename for the bedgraph file</span>
<span class="sd">        stream : bool</span>
<span class="sd">            Whether or not to read as stream</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicArray</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_dict({&#39;chr1&#39;: 30000, &#39;chr2&#39;: 31000, &#39;chr3&#39;: 32000})</span>
<span class="sd">        &gt;&gt;&gt; genome.read_track(&#39;example_data/small_treat_pileup.bdg&#39;)</span>
<span class="sd">        chr1: [ 0.0 0.0 1.0 ... 0.0 0.0 0.0]</span>
<span class="sd">        chr2: [ 0.0 0.0 0.0 ... 0.0 0.0 0.0]</span>
<span class="sd">        chr3: [ 0.0 0.0 0.0 ... 0.0 0.0 0.0]</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">content</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">stream</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_track</span><span class="p">(</span><span class="n">content</span><span class="p">)</span></div>


<div class="viewcode-block" id="Genome.get_intervals">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.get_intervals">[docs]</a>
    <span class="k">def</span> <span class="nf">get_intervals</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">intervals</span><span class="p">:</span> <span class="n">Interval</span><span class="p">,</span> <span class="n">stranded</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomicIntervals</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get genomic intervals from interval data. </span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        intervals : Interval</span>
<span class="sd">            Interval&#39;s or stream of Intervals&#39;s</span>
<span class="sd">        stranded : bool</span>
<span class="sd">            Wheter or not the intervals are stranded</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicIntervals</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; intervals = bnp.Interval(chromosome=[&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr2&#39;], start=[0, 10, 0], stop=[5, 15, 5])</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_dict({&#39;chr1&#39;: 20, &#39;chr2&#39;: 10})</span>
<span class="sd">        &gt;&gt;&gt; genome.get_intervals(intervals)</span>
<span class="sd">        Genomic Intervals on [&#39;chr1&#39;, &#39;chr2&#39;]:</span>
<span class="sd">        Interval with 3 entries</span>
<span class="sd">                       chromosome                    start                     stop</span>
<span class="sd">                             chr1                        0                        5</span>
<span class="sd">                             chr1                       10                       15</span>
<span class="sd">                             chr2                        0                        5</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">GenomicIntervals</span><span class="o">.</span><span class="n">from_intervals</span><span class="p">(</span><span class="n">intervals</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="p">,</span> <span class="n">is_stranded</span><span class="o">=</span><span class="n">stranded</span><span class="p">)</span></div>


<div class="viewcode-block" id="Genome.read_intervals">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.read_intervals">[docs]</a>
    <span class="k">def</span> <span class="nf">read_intervals</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">stranded</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">stream</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomicIntervals</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Read a bed file and represent it as `GenomicIntervals`</span>

<span class="sd">        If `stream` is `True` then read the bedgraph in chunks and get</span>
<span class="sd">        a lazy evaluated GenomicTrack</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            Filename for the bed file</span>
<span class="sd">        stranded : bool</span>
<span class="sd">            Whether or not to treat the intervals as stranded</span>
<span class="sd">        stream : bool</span>
<span class="sd">            Wheter to read as a stream</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicIntervals</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_file(&#39;example_data/small_sequence.fa&#39;)</span>
<span class="sd">        &gt;&gt;&gt; genome.read_intervals(&#39;example_data/small_summits.bed&#39;)</span>
<span class="sd">        Genomic Intervals on [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;]:</span>
<span class="sd">        Interval with 13 entries</span>
<span class="sd">                       chromosome                    start                     stop</span>
<span class="sd">                             chr1                      639                      640</span>
<span class="sd">                             chr1                     6023                     6024</span>
<span class="sd">                             chr1                     7124                     7125</span>
<span class="sd">                             chr2                      849                      850</span>
<span class="sd">                             chr2                     6320                     6321</span>
<span class="sd">                             chr2                     8483                     8484</span>
<span class="sd">                             chr2                    11342                    11343</span>
<span class="sd">                             chr2                    12527                    12528</span>
<span class="sd">                             chr2                    13092                    13093</span>
<span class="sd">                             chr2                    18943                    18944</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">path</span> <span class="o">=</span> <span class="n">PurePath</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="n">suffix</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">suffixes</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.gz&quot;</span><span class="p">:</span>
            <span class="n">suffix</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">suffixes</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">buffer_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">buffer_type</span> <span class="o">=</span> <span class="n">buffer_types</span><span class="p">[</span><span class="n">suffix</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">buffer_type</span> <span class="o">==</span> <span class="n">BedBuffer</span> <span class="ow">and</span> <span class="n">stranded</span><span class="p">:</span> 
                <span class="n">buffer_type</span> <span class="o">=</span> <span class="n">Bed6Buffer</span>
            <span class="k">if</span> <span class="n">buffer_type</span> <span class="o">==</span> <span class="n">BamBuffer</span><span class="p">:</span>
                <span class="n">buffer_type</span> <span class="o">=</span> <span class="n">BamIntervalBuffer</span>
        <span class="n">content</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">buffer_type</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_intervals</span><span class="p">(</span><span class="n">content</span><span class="p">,</span> <span class="n">stranded</span><span class="p">)</span></div>


<div class="viewcode-block" id="Genome.read_locations">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.read_locations">[docs]</a>
    <span class="k">def</span> <span class="nf">read_locations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">stranded</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">stream</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">has_numeric_chromosomes</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomicLocation</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Read a set of locations from file and convert them to `GenomicLocation`        </span>

<span class="sd">        Locations can be treated as stranded or not stranded. If `has_numeric_chromosomes` then &#39;chr&#39;</span>
<span class="sd">        will be prepended to all the chromosome names in order to make them compatible with the</span>
<span class="sd">        genome&#39;s chromosome names</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            Filename of the locations</span>
<span class="sd">        stranded : bool</span>
<span class="sd">            Whether or not the locations are stranded</span>
<span class="sd">        stream : bool</span>
<span class="sd">            Whether the data should be read as a stream</span>
<span class="sd">        has_numeric_chromosomes : bool</span>
<span class="sd">            Whether the file has the chromosomes as numbers</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicLocation</span>
<span class="sd">            (Stranded) GenomicLocation</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_file(&#39;example_data/hg38.chrom.sizes&#39;)</span>
<span class="sd">        &gt;&gt;&gt; genome.read_locations(&#39;example_data/thousand_genomes.vcf&#39;, has_numeric_chromosomes=True)</span>
<span class="sd">        Genomic Locations on [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;, &#39;chr4&#39;, &#39;chr5&#39;, &#39;chr6&#39;, &#39;chr7&#39;, &#39;chr8&#39;, &#39;chr9&#39;, &#39;chr10&#39;, &#39;...&#39;]:</span>
<span class="sd">        LocationEntry with 74 entries</span>
<span class="sd">                       chromosome                 position</span>
<span class="sd">                            chr21                  5033883</span>
<span class="sd">                            chr21                  5035657</span>
<span class="sd">                            chr21                  5038297</span>
<span class="sd">                            chr21                  5038312</span>
<span class="sd">                            chr21                  5052250</span>
<span class="sd">                            chr21                  5053935</span>
<span class="sd">                            chr21                  5053961</span>
<span class="sd">                            chr21                  5063903</span>
<span class="sd">                            chr21                  5063916</span>
<span class="sd">                            chr21                  5064678</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="ow">not</span> <span class="p">(</span><span class="n">stream</span> <span class="ow">and</span> <span class="n">has_numeric_chromosomes</span><span class="p">)</span>
        <span class="k">assert</span> <span class="ow">not</span> <span class="n">stranded</span><span class="p">,</span> <span class="s2">&quot;Stranded locations are not supported yet&quot;</span>
        <span class="n">f</span> <span class="o">=</span> <span class="n">bnp_open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">buffer_type</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">stream</span><span class="p">:</span>
            <span class="n">data_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">data_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="n">LocationEntry</span><span class="o">.</span><span class="n">empty</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">data_list</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_locations</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">has_numeric_chromosomes</span><span class="o">=</span><span class="n">has_numeric_chromosomes</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">_mask_data_on_extra_chromosomes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">chromosome_field_name</span><span class="o">=</span><span class="s1">&#39;chromosome&#39;</span><span class="p">):</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">BNPDataClass</span><span class="p">))</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">data</span>
        <span class="n">encoded_chromosomes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">encoding</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">chromosome_field_name</span><span class="p">))</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">replace</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="o">**</span><span class="p">{</span><span class="n">chromosome_field_name</span><span class="p">:</span> <span class="n">encoded_chromosomes</span><span class="p">})</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">is_included</span><span class="p">(</span><span class="n">encoded_chromosomes</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>

<div class="viewcode-block" id="Genome.get_locations">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.get_locations">[docs]</a>
    <span class="k">def</span> <span class="nf">get_locations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">LocationEntry</span><span class="p">,</span> <span class="n">has_numeric_chromosomes</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomicLocation</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create `GenomicLocation`s from location entries</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        data : LocationEntry</span>
<span class="sd">            BNPDataClass of location entries</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicLocation</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_file(&#39;example_data/small_sequence.fa&#39;)</span>
<span class="sd">        &gt;&gt;&gt; genome.get_locations(bnp.datatypes.LocationEntry(chromosome=[&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr2&#39;], position=[0, 10, 0]))</span>
<span class="sd">        Genomic Locations on [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;]:</span>
<span class="sd">        LocationEntry with 3 entries</span>
<span class="sd">                       chromosome                 position</span>
<span class="sd">                             chr1                        0</span>
<span class="sd">                             chr1                       10</span>
<span class="sd">                             chr2                        0</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">has_numeric_chromosomes</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">replace</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="n">as_encoded_array</span><span class="p">([</span><span class="s1">&#39;chr&#39;</span><span class="o">+</span><span class="n">chromosome</span><span class="o">.</span><span class="n">to_string</span><span class="p">()</span> <span class="k">for</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">chromosome</span><span class="p">]))</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mask_data_on_extra_chromosomes</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">GenomicLocation</span><span class="o">.</span><span class="n">from_data</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="p">)</span></div>


<div class="viewcode-block" id="Genome.read_sequence">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.read_sequence">[docs]</a>
    <span class="k">def</span> <span class="nf">read_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomicSequence</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Read the genomic sequence from file.</span>

<span class="sd">        If a `fasta` file was used to create the Genome object, `filename` can be `None`</span>
<span class="sd">        in which case the sequence will be read from that fasta file</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicSequence</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_file(&#39;example_data/small_sequence.fa&#39;)</span>
<span class="sd">        &gt;&gt;&gt; genome.read_sequence()</span>
<span class="sd">        GenomicSequence over chromosomes: [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;]</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_file(&#39;example_data/small.chrom.sizes&#39;)</span>
<span class="sd">        &gt;&gt;&gt; genome.read_sequence(&#39;example_data/small_sequence.fa&#39;)</span>
<span class="sd">        GenomicSequence over chromosomes: [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fasta_filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fasta_filename</span>
        <span class="k">return</span> <span class="n">GenomicSequence</span><span class="o">.</span><span class="n">from_indexed_fasta</span><span class="p">(</span><span class="n">open_indexed</span><span class="p">(</span><span class="n">filename</span><span class="p">),</span> <span class="n">genome_context</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="p">)</span></div>


<div class="viewcode-block" id="Genome.read_annotation">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.read_annotation">[docs]</a>
    <span class="k">def</span> <span class="nf">read_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomicAnnotation</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Read genomic annotation from a &#39;gtf&#39; or &#39;gff&#39; file</span>

<span class="sd">        The annotation file should have &#39;gene&#39;, &#39;transcript&#39; and &#39;exon&#39; entries.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">            Filename of the annotation file</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicAnnotation</span>
<span class="sd">            `GenomicAnnotation` containing the genes, transcripts and exons</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">        &gt;&gt;&gt; genome = bnp.Genome.from_file(&#39;example_data/small_sequence.fa&#39;)</span>
<span class="sd">        &gt;&gt;&gt; genome.read_annotation(&#39;example_data/small.gtf&#39;)</span>
<span class="sd">        GenomicAnnotation(genome_context=[&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;], data=GTFEntry with 5 entries</span>
<span class="sd">                       chromosome                   source             feature_type                    start                     stop                    score                   strand                    phase                atributes</span>
<span class="sd">                             chr1                knownGene               transcript                    17369                    17436                        .                        -                        .  gene_id &quot;ENST0000061921</span>
<span class="sd">                             chr1                knownGene                     exon                    17369                    17436                        .                        -                        .  gene_id &quot;ENST0000061921</span>
<span class="sd">                             chr1                knownGene               transcript                    29554                    31097                        .                        +                        .  gene_id &quot;ENST0000047335</span>
<span class="sd">                             chr1                knownGene                     exon                    29554                    30039                        .                        +                        .  gene_id &quot;ENST0000047335</span>
<span class="sd">                             chr1                knownGene                     exon                    30564                    30667                        .                        +                        .  gene_id &quot;ENST0000047335)</span>


<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="n">gtf_entries</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">stream</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">GenomicAnnotation</span><span class="o">.</span><span class="n">from_gtf_entries</span><span class="p">(</span><span class="n">gtf_entries</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">(&quot;</span> <span class="o">+</span> <span class="nb">repr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;)&quot;</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">table</span><span class="p">(</span>
            <span class="p">((</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">chrom_sizes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="s1">&#39;_&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">key</span><span class="p">),</span>
            <span class="n">headers</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;Size&quot;</span><span class="p">])</span>

<div class="viewcode-block" id="Genome.get_genome_context">
<a class="viewcode-back" href="../../../modules/genomic_data.html#bionumpy.genomic_data.Genome.get_genome_context">[docs]</a>
    <span class="k">def</span> <span class="nf">get_genome_context</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenomeContext</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Get the genome context of the Genome</span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomeContext</span>
<span class="sd">        &#39;&#39;&#39;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span></div>


    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">size</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;The size of the genome&#39;&#39;&#39;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_context</span><span class="o">.</span><span class="n">size</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>