

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Working with Multiple Files/Data Sources &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Broadcastable Functions" href="broadcastable_functions.html" />
    <link rel="prev" title="Intervals" href="intervals.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="intervals.html">Intervals</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Working with Multiple Files/Data Sources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sort-order">Sort Order</a></li>
<li class="toctree-l2"><a class="reference internal" href="#two-indexed-files">Two Indexed Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Working with Multiple Files/Data Sources</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/source/multiple_data_sources.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="working-with-multiple-files-data-sources">
<span id="multiple-data-sources"></span><h1>Working with Multiple Files/Data Sources<a class="headerlink" href="#working-with-multiple-files-data-sources" title="Link to this heading">ÔÉÅ</a></h1>
<p>A lot of analyses requires working on multiple data sources that are annotated on the same reference sequences. Especially when working with reference genomes. A common example is analysing some variants from a vcf file, but only those variants that fall within some intervals specified in a bed file. One way to do this in BioNumPy is to use the <cite>MultiStream</cite> class. This is designed to take in multiple data streams or indexed data and deliver synchronized streams that return data corresponding to the same reference sequence/chromosome in each chunk. It is best shown in an example.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/few_variants.vcf&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intervals</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/small_interval.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open_indexed</span><span class="p">(</span><span class="s2">&quot;example_data/small_genome.fa&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>These variants and intervals come in chunks, and stream of chunks are not really satisfied. In addition the reference genome is read as an indexed file, and so is able to give us arbitrary parts of the reference genome at demand. We can synch these three sources up using MultiStream:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">multistream</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">MultiStream</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">get_contig_lengths</span><span class="p">(),</span>
<span class="gp">... </span>                              <span class="n">sequence</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">intervals</span><span class="o">=</span><span class="n">intervals</span><span class="p">)</span>
</pre></div>
</div>
<p>The attributes we specify for the multistream are now synched streams that give data corresponding to the sequences listed in <cite>reference.get_sequence_lengths()</cite> one at a time. This means we can give these streams to any function with the <cite>streamable</cite> decorator. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.arithmetics</span> <span class="kn">import</span> <span class="n">get_boolean_mask</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nd">@bnp</span><span class="o">.</span><span class="n">streamable</span><span class="p">(</span><span class="nb">sum</span><span class="p">)</span>
<span class="gp">... </span><span class="k">def</span> <span class="nf">get_letter_counts_around_variants</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">,</span> <span class="n">variants</span><span class="p">,</span> <span class="n">intervals</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">mask</span> <span class="o">=</span> <span class="n">get_boolean_mask</span><span class="p">(</span><span class="n">intervals</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">))</span>
<span class="gp">... </span>    <span class="n">positions</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">position</span>
<span class="gp">... </span>    <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">mask</span><span class="p">[</span><span class="n">positions</span><span class="p">]]</span> <span class="c1"># Only look at variants inside intervals</span>
<span class="gp">... </span>    <span class="n">surrounding_sequences</span> <span class="o">=</span> <span class="n">reference_sequence</span><span class="p">[</span><span class="n">positions</span><span class="o">-</span><span class="mi">10</span><span class="p">:</span><span class="n">positions</span><span class="o">+</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_encoded</span><span class="p">(</span><span class="n">bnp</span><span class="o">.</span><span class="n">as_encoded_array</span><span class="p">(</span><span class="n">surrounding_sequences</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">))</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">get_letter_counts_around_variants</span><span class="p">(</span><span class="n">multistream</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">multistream</span><span class="o">.</span><span class="n">variants</span><span class="p">,</span> <span class="n">multistream</span><span class="o">.</span><span class="n">intervals</span><span class="p">)</span>
<span class="go">EncodedCounts(alphabet=[&#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;], counts=array([155, 127, 116, 122]), row_names=None)</span>
</pre></div>
</div>
<section id="sort-order">
<h2>Sort Order<a class="headerlink" href="#sort-order" title="Link to this heading">ÔÉÅ</a></h2>
<p>Unindexed files (.bed, .vcf) are read from start to finish. This means that the data for each chromosome comes one at a time in the order they appear in the file. Multistream expects that this order is the same as that specified in the <cite>seqeuence_lengths</cite> parameter, which is unfortunately not always the case. This is usually due to the fact that different programs sorts the files in different ways: <cite>[‚Äúchr1‚Äù, ‚Äúchr2‚Äù,,, ‚Äúchr10‚Äù, ‚Äúchr11‚Äù]</cite> or <cite>[‚Äúchr1‚Äù, ‚Äúchr10‚Äù, ‚Äúchr11‚Äù,,, ‚Äúchr2‚Äù]</cite>. If you only have one unindexed file, the easiest solution to a sort order discrepancy is to change the sort order of the <cite>sequence_lengths</cite> parameter. This can be done with the <cite>sort_dict_by_key</cite> function, with <cite>human_key_func</cite> or <cite>None</cite> as key.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence_lengths</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;chr1&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="s2">&quot;chr11&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">:</span> <span class="mi">40</span><span class="p">,</span> <span class="s2">&quot;chr10&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bnp</span><span class="o">.</span><span class="n">MultiStream</span><span class="o">.</span><span class="n">sort_dict_by_key</span><span class="p">(</span><span class="n">sequence_lengths</span><span class="p">)</span>
<span class="go">{&#39;chr1&#39;: 10, &#39;chr10&#39;: 50, &#39;chr11&#39;: 20, &#39;chr2&#39;: 40}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bnp</span><span class="o">.</span><span class="n">MultiStream</span><span class="o">.</span><span class="n">sort_dict_by_key</span><span class="p">(</span><span class="n">sequence_lengths</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">bnp</span><span class="o">.</span><span class="n">MultiStream</span><span class="o">.</span><span class="n">human_key_func</span><span class="p">)</span>
<span class="go">{&#39;chr1&#39;: 10, &#39;chr2&#39;: 40, &#39;chr10&#39;: 50, &#39;chr11&#39;: 20}</span>
</pre></div>
</div>
</section>
<section id="two-indexed-files">
<h2>Two Indexed Files<a class="headerlink" href="#two-indexed-files" title="Link to this heading">ÔÉÅ</a></h2>
<p>If you have two unindexed files, with conflicting sort order, it is not enough to change the sort order of the <cite>sequence_lenghts</cite>. Hopefully, one of the files is small enough that it can fit into memory, so that we can turn it into a dict with chromosomes as key. Such a dict can be passed into the MultiStream, and is oblivious to the sort-order of the origin file. For instance, a bed file with intervals is usually quite small (unless it represents mapped reads):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">intervals</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/small_interval.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">bnp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">intervals</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval_dict</span>
<span class="go">{&#39;0&#39;: Interval with 5 entries</span>
<span class="go">               chromosome                    start                     stop</span>
<span class="go">                        0                       13                       18</span>
<span class="go">                        0                       37                       46</span>
<span class="go">                        0                       62                       83</span>
<span class="go">                        0                      105                      126</span>
<span class="go">                        0                      129                      130, &#39;1&#39;: Interval with 10 entries</span>
<span class="go">               chromosome                    start                     stop</span>
<span class="go">                        1                        3                       21</span>
<span class="go">                        1                       41                       65</span>
<span class="go">                        1                       91                      114</span>
<span class="go">                        1                      131                      153</span>
<span class="go">                        1                      157                      168</span>
<span class="go">                        1                      174                      201</span>
<span class="go">                        1                      213                      230</span>
<span class="go">                        1                      240                      268</span>
<span class="go">                        1                      290                      315</span>
<span class="go">                        1                      319                      339, &#39;2&#39;: Interval with 15 entries</span>
<span class="go">               chromosome                    start                     stop</span>
<span class="go">                        2                        2                       16</span>
<span class="go">                        2                       44                       49</span>
<span class="go">                        2                       77                      101</span>
<span class="go">                        2                      108                      127</span>
<span class="go">                        2                      135                      154</span>
<span class="go">                        2                      163                      165</span>
<span class="go">                        2                      173                      177</span>
<span class="go">                        2                      201                      214</span>
<span class="go">                        2                      242                      268</span>
<span class="go">                        2                      292                      320, &#39;3&#39;: Interval with 20 entries</span>
<span class="go">               chromosome                    start                     stop</span>
<span class="go">                        3                        7                       34</span>
<span class="go">                        3                       58                       82</span>
<span class="go">                        3                       95                      101</span>
<span class="go">                        3                      130                      138</span>
<span class="go">                        3                      150                      170</span>
<span class="go">                        3                      188                      211</span>
<span class="go">                        3                      234                      261</span>
<span class="go">                        3                      283                      302</span>
<span class="go">                        3                      325                      352</span>
<span class="go">                        3                      353                      362}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">multistream</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">MultiStream</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">get_contig_lengths</span><span class="p">(),</span>
<span class="gp">... </span>                              <span class="n">sequence</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">intervals</span><span class="o">=</span><span class="n">interval_dict</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="intervals.html" class="btn btn-neutral float-left" title="Intervals" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="broadcastable_functions.html" class="btn btn-neutral float-right" title="Broadcastable Functions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>