<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Working with Multiple Files/Data Sources &mdash; bionumpy  documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Broadcastable Functions" href="broadcastable_functions.html" />
    <link rel="prev" title="Intervals" href="intervals.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../readme.html">What is BioNumPy?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation and getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#your-first-bionumpy-program">Your first BioNumPy program</a></li>
</ul>
<p class="caption"><span class="caption-text">BioNumPy Concepts</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="working_with_big_data.html">Working with big data</a></li>
<li class="toctree-l1"><a class="reference internal" href="supported_file_formats.html">Currently supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="supported_file_formats.html#implementing-a-new-file-format">Implementing a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="intervals.html">Intervals</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="broadcastable_functions.html">Broadcastable Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="rollable_functions.html">Rollable Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarization.html">Summarization</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/bam_handling.html">BAM Handling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastq_filtering.html">FASTQ filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastqc.html">Simple quality checking of fastq files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/numpy_operations.html">Using Numpy Functionality</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
</ul>
<p class="caption"><span class="caption-text">Developer Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/getting_started.html">Getting started and how to contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/setting_up_development_environment.html">Setting up your development environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/making_examples.html">Making examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/writing_documentation.html">Writing documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/design_principles.html">Design principles and code overview</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Working with Multiple Files/Data Sources</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/source/multiple_data_sources.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="working-with-multiple-files-data-sources">
<span id="intervals"></span><h1>Working with Multiple Files/Data Sources<a class="headerlink" href="#working-with-multiple-files-data-sources" title="Permalink to this headline">Â¶</a></h1>
<p>A lot of analyses requires working on multiple data sources that are annotated on the same reference sequences. Especially when workin with reference genomes. A common example is analysing soms variants from a vcf file, but only those variants that fall within some intervals specified in a bed file. One way to do this in BioNumPy is to use the <cite>MultiStream</cite> class. This is designed to take in multiple data streams or indexed data and deliver synchronized streams that return data corresponding to the same reference sequence/chromosome in each chunk. It is best shown in an example.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/few_variants.vcf&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intervals</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/small_interval.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open_indexed</span><span class="p">(</span><span class="s2">&quot;example_data/small_genome.fa&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>These variants and intervals come in chunks, and stream of chunks are not really satisfied. In addition the reference genome is read as an indexed file, and so is able to give us arbitrary parts of the reference genome at demand. We can synch these three sources up using MultiStream:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">multistream</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">MultiStream</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">get_contig_lengths</span><span class="p">(),</span>
<span class="gp">... </span>                              <span class="n">sequence</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">intervals</span><span class="o">=</span><span class="n">intervals</span><span class="p">)</span>
</pre></div>
</div>
<p>The attributes we specify for the multistream are now synched streams that give data corresponding to the seqeunces listed int <cite>reference.get_sequence_lengths()</cite> one at a time. This means we can give these streams to any function with the <cite>streamable</cite> decorator. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.intervals</span> <span class="k">import</span> <span class="n">get_boolean_mask</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nd">@bnp</span><span class="o">.</span><span class="n">streamable</span><span class="p">(</span><span class="nb">sum</span><span class="p">)</span>
<span class="gp">... </span><span class="k">def</span> <span class="nf">get_letter_counts_around_variants</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">,</span> <span class="n">variants</span><span class="p">,</span> <span class="n">intervals</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">mask</span> <span class="o">=</span> <span class="n">get_boolean_mask</span><span class="p">(</span><span class="n">intervals</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">reference_sequence</span><span class="p">))</span>
<span class="gp">... </span>    <span class="n">positions</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">position</span>
<span class="gp">... </span>    <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">mask</span><span class="p">[</span><span class="n">positions</span><span class="p">]]</span> <span class="c1"># Only look at variants inside intervals</span>
<span class="gp">... </span>    <span class="n">surrounding_sequences</span> <span class="o">=</span> <span class="n">reference_sequence</span><span class="p">[</span><span class="n">positions</span><span class="o">-</span><span class="mi">10</span><span class="p">:</span><span class="n">positions</span><span class="o">+</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_encoded</span><span class="p">(</span><span class="n">bnp</span><span class="o">.</span><span class="n">as_encoded_array</span><span class="p">(</span><span class="n">surrounding_sequences</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">))</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">get_letter_counts_around_variants</span><span class="p">(</span><span class="n">multistream</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">multistream</span><span class="o">.</span><span class="n">variants</span><span class="p">,</span> <span class="n">multistream</span><span class="o">.</span><span class="n">intervals</span><span class="p">)</span>
<span class="go">EncodedCounts(alphabet=[&#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;], counts=array([155, 127, 116, 122]), row_names=None)</span>
</pre></div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="intervals.html" class="btn btn-neutral float-left" title="Intervals" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="broadcastable_functions.html" class="btn btn-neutral float-right" title="Broadcastable Functions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>