<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Reading files &mdash; bionumpy  documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Working with big data" href="working_with_big_data.html" />
    <link rel="prev" title="Installation and getting started" href="../installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../readme.html">What is BioNumPy?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation and getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#your-first-bionumpy-program">Your first BioNumPy program</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Reading files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#method-1-reading-all-your-data-at-once">Method 1: Reading all your data at once</a></li>
<li class="toctree-l2"><a class="reference internal" href="#method-2-reading-a-single-chunk-from-a-large-file">Method 2: Reading a single chunk from a large file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#method-3-reading-whole-file-as-a-stream-of-chunks">Method 3: Reading whole file as a stream of chunks</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="working_with_big_data.html">Working with big data</a></li>
<li class="toctree-l1"><a class="reference internal" href="supported_file_formats.html">Currently supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="supported_file_formats.html#implementing-a-new-file-format">Implementing a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="broadcastable_functions.html">Broadcastable Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="rollable_functions.html">Rollable Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarization.html">Summarization</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/bam_handling.html">BAM Handling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastq_filtering.html">FASTQ filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastqc.html">Simple quality checking of fastq files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/numpy_operations.html">Using Numpy Functionality</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Reading files</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/source/reading_files.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="reading-files">
<span id="id1"></span><h1>Reading files<a class="headerlink" href="#reading-files" title="Permalink to this headline">¶</a></h1>
<p>There are three main ways of getting your data into memory in BioNumPy. Which way you choose mainly depends on how big your data is:</p>
<div class="section" id="method-1-reading-all-your-data-at-once">
<h2>Method 1: Reading all your data at once<a class="headerlink" href="#method-1-reading-all-your-data-at-once" title="Permalink to this headline">¶</a></h2>
<p>If you are working with data sets that are small enough to fit into memory, you can read the whole file. The benefit of this approach is that you don’t need to take into account that the data has been split into chunks (next methods):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;filename&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<p>In the above example, <cite>data</cite> will be an NpDataClass object with various fields, depending on the file type. For instance, if you read a fastq file, you will be able to access the sequences (<cite>data.sequence</cite>) and the base qualities (<cite>data.quality</cite>).</p>
</div>
<div class="section" id="method-2-reading-a-single-chunk-from-a-large-file">
<h2>Method 2: Reading a single chunk from a large file<a class="headerlink" href="#method-2-reading-a-single-chunk-from-a-large-file" title="Permalink to this headline">¶</a></h2>
<p>If you have a large file, and don’t want to read all of it, you can read a single chunk. This can be useful if you only want to get to know your data or perform an analysis on a small subset of your data:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;reads.fastq&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chunk</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">read_chunk</span><span class="p">(</span><span class="n">chunk_size</span><span class="o">=</span><span class="mi">10000000</span><span class="p">)</span>
</pre></div>
</div>
<p>Here <cite>chunk_size</cite> is the number of bytes to read. If the file is small enough, you may get the whole file by setting chunk size to a big number, but if you want to make sure you read the whole file as one single cunk you should choose method 1 instead.</p>
</div>
<div class="section" id="method-3-reading-whole-file-as-a-stream-of-chunks">
<h2>Method 3: Reading whole file as a stream of chunks<a class="headerlink" href="#method-3-reading-whole-file-as-a-stream-of-chunks" title="Permalink to this headline">¶</a></h2>
<p>This is the preferred way you should use for all large files (typically fasta files, fastq files and bam files). The idea is to read the whole file as chunks that can be iterated over. This way, only one single chunk will be kept in memory at the time.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;file.fastq&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
<span class="gp">&gt;&gt;&gt; </span>   <span class="c1"># do stuff with this chunk, e.g. take the man</span>
<span class="gp">&gt;&gt;&gt; </span>   <span class="c1"># of the base qualities</span>
</pre></div>
</div>
<p>Remember that when one chunk has been processed, that chunk is “lost” and the generator continues to the next chunk. A typical implementation will thus be to write a function that does all you want to do one one chunk, run that function on each chunk and summarize the results in some way.</p>
<p>Continue to the guide on <a class="reference internal" href="working_with_big_data.html#working-with-big-data"><span class="std std-ref">Working with big data</span></a> to see the recommended way of working with chunks in BioNumPy.</p>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../installation.html" class="btn btn-neutral float-left" title="Installation and getting started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="working_with_big_data.html" class="btn btn-neutral float-right" title="Working with big data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>