

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Supported file formats &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Sequences" href="sequences.html" />
    <link rel="prev" title="Encodings" href="encoding.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="encoding.html">Encodings</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Supported file formats</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Supported file formats</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/source/supported_file_formats.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="supported-file-formats">
<span id="id1"></span><h1>Supported file formats<a class="headerlink" href="#supported-file-formats" title="Link to this heading"></a></h1>
<p>This is a list of  currently supported file formats in BioNumPy. Reading files with any of these extensions with <cite>bnp.open</cite> will make BioNumPy automatically detect the file type and read the data into an appropriate data structure (which will be a dataclass-like object with fields).</p>
<ul class="simple">
<li><p>vcf</p></li>
<li><p>bed</p></li>
<li><p>fasta / fa</p></li>
<li><p>bed</p></li>
<li><p>fastq / fq</p></li>
<li><p>gfa (limited support only)</p></li>
<li><p>gff</p></li>
<li><p>gtf</p></li>
<li><p>gff3</p></li>
<li><p>sam / bam</p></li>
</ul>
<section id="example">
<h2>Example<a class="headerlink" href="#example" title="Link to this heading"></a></h2>
<p>We open a bed file, read one chunk and print a description of that chunk:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="kn">from</span> <span class="nn">bionumpy.io.delimited_buffers</span> <span class="kn">import</span> <span class="n">DelimitedBuffer</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/test.bed&quot;</span><span class="p">)</span>
<span class="n">chunk</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">read_chunk</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
</pre></div>
</div>
<p>The above example should work wih any of the supported file formats.</p>
<p>This shows us that we have a a chunk of 71 intervals, and we get to see the first of these:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Interval with 71 entries
               chromosome                    start                     stop
                       17                  7512371                  7512447
                       17                  7512377                  7512453
                       17                  7512393                  7512469
                       17                  7512420                  7512496
                       17                  7512422                  7512473
                       17                  7512425                  7512501
                       17                  7512428                  7512504
                       17                  7512474                  7512550
                       17                  7512537                  7512613
                       17                  7512559                  7512635
</pre></div>
</div>
</section>
</section>
<section id="reading-a-new-file-format">
<h1>Reading a new file format<a class="headerlink" href="#reading-a-new-file-format" title="Link to this heading"></a></h1>
<p>BioNumpy works well with popular file formats in biological data.
However, if you have a custom file format that you would like to read into BioNumpy,
you can implement a new buffer class that inherits from <cite>DelimitedBuffer</cite> and
specify the dataclass that you would like to use to store the data.
Here is an example of how you can implement a new buffer class for a custom file format:</p>
<p>We define a custom dataclass (e.g. MyCustomFormat here) that corresponds to the columns in our file format.
We then define a new buffer class (e.g. MyCustomBuffer here) that inherits from <cite>DelimitedBuffer</cite> and specify
the dataclass (e.g. MyCustomFormat here) that we would like to use.
We can then use the <cite>bnp.open</cite> function to read all the files that have similar format.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">bionumpy.io.delimited_buffers</span> <span class="kn">import</span> <span class="n">DelimitedBuffer</span>
<span class="kn">from</span> <span class="nn">bionumpy.bnpdataclass</span> <span class="kn">import</span> <span class="n">bnpdataclass</span>
<span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>

<span class="nd">@bnpdataclass</span>
<span class="k">class</span> <span class="nc">MyCustomFormat</span><span class="p">:</span>
  <span class="n">dna</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span>
  <span class="n">amino_acid</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">AminoAcidEncoding</span>
  <span class="n">v_gene</span><span class="p">:</span> <span class="nb">str</span>
  <span class="n">j_gene</span><span class="p">:</span> <span class="nb">str</span>

<span class="k">class</span> <span class="nc">MyCustomBuffer</span><span class="p">(</span><span class="n">DelimitedBuffer</span><span class="p">):</span>
   <span class="n">dataclass</span> <span class="o">=</span> <span class="n">MyCustomFormat</span>

<span class="n">my_sequence_data</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;example_data/airr.tsv&quot;</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">MyCustomBuffer</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">my_sequence_data</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>MyCustomFormat with 100 entries
                         dna               amino_acid                   v_gene                   j_gene
     TGCGCCACCTGGGGGGACGAGCA               CATWGDEQYF                 TRBV10-2                  TRBJ2-7
     TGTGCCAGCTCACCTACGAATTC         CASSPTNSGSNYGYTF                   TRBV18                  TRBJ1-2
     TGCGGGCCCGTAATGAACACTGA              CGPVMNTEAFF                 TRBV10-2                  TRBJ1-1
     TGTGCCAGCAGTGAAGCGCGTCC         CASSEARPARMYGYTF                  TRBV6-1                  TRBJ1-2
     TGTGCCAGCAGTAGTGGGACAGG          CASSSGTGPDQPQHF                  TRBV6-3                  TRBJ1-5
     TGTGCCAGCAACCTAGCGGGGAA          CASNLAGKNTGELFF                  TRBV6-2                  TRBJ2-2
     TGTGCCAGCAGCCAACCGGGGGG         CASSQPGGSGNYGYTF                  TRBV4-2                  TRBJ1-2
     TGCGCCAGCAGCCGCGGCCTCAG           CASSRGLREETQYF                  TRBV5-1                  TRBJ2-5
     TGTGCCAGCAGCCAAGTCTCACG        CASSQVSRQDSSYEQYF                  TRBV4-2                  TRBJ2-7
     TGTGCCAGCAGGCCGGGACAGGG     CASRPGQGAPGWEDNYGYTF                   TRBV28                  TRBJ1-2
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="encoding.html" class="btn btn-neutral float-left" title="Encodings" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="sequences.html" class="btn btn-neutral float-right" title="Sequences" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>