

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bionumpy.io.files &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../menu.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bionumpy.io.files</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bionumpy.io.files</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">PurePath</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Optional</span>

<span class="kn">from</span> <span class="nn">.gzip_reading</span> <span class="kn">import</span> <span class="n">gzip</span>
<span class="c1"># import gzip</span>
<span class="kn">import</span> <span class="nn">dataclasses</span>
<span class="kn">from</span> <span class="nn">.file_buffers</span> <span class="kn">import</span> <span class="n">FileBuffer</span>
<span class="kn">from</span> <span class="nn">.fastq_buffer</span> <span class="kn">import</span> <span class="n">FastQBuffer</span>
<span class="kn">from</span> <span class="nn">.multiline_buffer</span> <span class="kn">import</span> <span class="n">MultiLineFastaBuffer</span>
<span class="kn">from</span> <span class="nn">.bam</span> <span class="kn">import</span> <span class="n">BamBuffer</span><span class="p">,</span> <span class="n">BamIntervalBuffer</span>
<span class="kn">from</span> <span class="nn">.delimited_buffers</span> <span class="kn">import</span> <span class="p">(</span><span class="n">BedBuffer</span><span class="p">,</span> <span class="n">GfaSequenceBuffer</span><span class="p">,</span>
                                <span class="n">GFFBuffer</span><span class="p">,</span> <span class="n">ChromosomeSizeBuffer</span><span class="p">,</span>
                                <span class="n">NarrowPeakBuffer</span><span class="p">,</span> <span class="n">BdgBuffer</span><span class="p">,</span> <span class="n">GTFBuffer</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">.buffers.sam</span> <span class="kn">import</span> <span class="n">SAMBuffer</span>
<span class="kn">from</span> <span class="nn">.vcf_buffers</span> <span class="kn">import</span> <span class="n">VCFBuffer</span>
<span class="kn">from</span> <span class="nn">.pairs</span> <span class="kn">import</span> <span class="n">PairsBuffer</span>
<span class="kn">from</span> <span class="nn">.wig</span> <span class="kn">import</span> <span class="n">WigBuffer</span>
<span class="kn">from</span> <span class="nn">.parser</span> <span class="kn">import</span> <span class="n">NumpyFileReader</span><span class="p">,</span> <span class="n">NpBufferedWriter</span><span class="p">,</span> <span class="n">NumpyBamWriter</span>
<span class="kn">from</span> <span class="nn">.exceptions</span> <span class="kn">import</span> <span class="n">FormatException</span>
<span class="kn">from</span> <span class="nn">..streams</span> <span class="kn">import</span> <span class="n">NpDataclassStream</span>
<span class="kn">from</span> <span class="nn">..bnpdataclass</span> <span class="kn">import</span> <span class="n">BNPDataClass</span><span class="p">,</span> <span class="n">bnpdataclass</span>
<span class="kn">from</span> <span class="nn">.npdataclassreader</span> <span class="kn">import</span> <span class="n">NpDataclassReader</span>
<span class="kn">import</span> <span class="nn">logging</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<span class="n">buffer_types</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;.vcf&quot;</span><span class="p">:</span> <span class="n">VCFBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.bed&quot;</span><span class="p">:</span> <span class="n">BedBuffer</span><span class="p">,</span>
    <span class="s1">&#39;.bdg&#39;</span><span class="p">:</span> <span class="n">BdgBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.narrowPeak&quot;</span><span class="p">:</span> <span class="n">NarrowPeakBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.fasta&quot;</span><span class="p">:</span> <span class="n">MultiLineFastaBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.fa&quot;</span><span class="p">:</span> <span class="n">MultiLineFastaBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.fna&quot;</span><span class="p">:</span> <span class="n">MultiLineFastaBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.faa&quot;</span><span class="p">:</span> <span class="n">MultiLineFastaBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.fastq&quot;</span><span class="p">:</span> <span class="n">FastQBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.fq&quot;</span><span class="p">:</span> <span class="n">FastQBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.gfa&quot;</span><span class="p">:</span> <span class="n">GfaSequenceBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.gff&quot;</span><span class="p">:</span> <span class="n">GFFBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.gtf&quot;</span><span class="p">:</span> <span class="n">GTFBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.gff3&quot;</span><span class="p">:</span> <span class="n">GFFBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.sam&quot;</span><span class="p">:</span> <span class="n">SAMBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.bam&quot;</span><span class="p">:</span> <span class="n">BamBuffer</span><span class="p">,</span>
    <span class="s2">&quot;.sizes&quot;</span><span class="p">:</span> <span class="n">ChromosomeSizeBuffer</span><span class="p">,</span>
    <span class="s1">&#39;.wig&#39;</span><span class="p">:</span> <span class="n">WigBuffer</span><span class="p">,</span>
    <span class="s1">&#39;.pairs&#39;</span><span class="p">:</span> <span class="n">PairsBuffer</span><span class="p">,</span>
    <span class="s1">&#39;.pa5&#39;</span><span class="p">:</span> <span class="n">PairsBuffer</span><span class="p">,</span>
<span class="p">}</span>


<span class="k">def</span> <span class="nf">_get_buffered_file</span><span class="p">(</span>
    <span class="n">filename</span><span class="p">,</span> <span class="n">suffix</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">is_gzip</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span>
<span class="p">):</span>
    <span class="n">open_func</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span> <span class="k">if</span> <span class="n">is_gzip</span> <span class="k">else</span> <span class="nb">open</span>

    <span class="k">if</span> <span class="n">buffer_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">buffer_type</span> <span class="o">=</span> <span class="n">_get_buffer_type</span><span class="p">(</span><span class="n">suffix</span><span class="p">)</span>
    <span class="n">writer_class</span> <span class="o">=</span> <span class="n">NpBufferedWriter</span>
    <span class="k">if</span> <span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.bam&quot;</span><span class="p">:</span>
        <span class="n">writer_class</span> <span class="o">=</span> <span class="n">NumpyBamWriter</span>
    <span class="k">if</span> <span class="n">mode</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="s2">&quot;write&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">writer_class</span><span class="p">(</span><span class="n">open_func</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">),</span> <span class="n">buffer_type</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">mode</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;append&#39;</span><span class="p">,</span> <span class="s1">&#39;ab&#39;</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">writer_class</span><span class="p">(</span><span class="n">open_func</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;ab&#39;</span><span class="p">),</span> <span class="n">buffer_type</span><span class="p">)</span>

    <span class="c1"># kwargs2 = {key: val for key, val in kwargs.items() if key in [&quot;has_header&quot;]}</span>
    <span class="n">file_reader</span> <span class="o">=</span> <span class="n">NumpyFileReader</span><span class="p">(</span><span class="n">open_func</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">),</span> <span class="n">buffer_type</span><span class="p">)</span> <span class="c1"># , **kwargs2)</span>
    <span class="k">if</span> <span class="n">is_gzip</span><span class="p">:</span>
        <span class="n">file_reader</span><span class="o">.</span><span class="n">set_prepend_mode</span><span class="p">()</span>
    <span class="n">lazy</span><span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;lazy&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">NpDataclassReader</span><span class="p">(</span><span class="n">file_reader</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="n">lazy</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_get_buffer_type</span><span class="p">(</span><span class="n">suffix</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">suffix</span> <span class="ow">in</span> <span class="n">buffer_types</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">buffer_types</span><span class="p">[</span><span class="n">suffix</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">all_buffer_types</span> <span class="o">=</span> <span class="p">{</span><span class="n">buffer</span><span class="o">.</span><span class="vm">__name__</span> <span class="k">for</span> <span class="n">buffer</span> <span class="ow">in</span> <span class="n">buffer_types</span><span class="o">.</span><span class="n">values</span><span class="p">()}</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File format </span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2"> does not have a default buffer type. &quot;</span>
                           <span class="sa">f</span><span class="s2">&quot;Specify buffer_type argument using  &quot;</span>
                           <span class="sa">f</span><span class="s2">&quot;one of </span><span class="si">{</span><span class="nb">str</span><span class="p">([</span><span class="s1">&#39;bnp.&#39;</span><span class="o">+</span><span class="n">b</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">b</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">all_buffer_types</span><span class="p">])</span><span class="si">}</span><span class="s2"> or change to a valid extension: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">buffer_types</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="bnp_open">
<a class="viewcode-back" href="../../../modules/io.html#bionumpy.io.bnp_open">[docs]</a>
<span class="k">def</span> <span class="nf">bnp_open</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">mode</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">NpDataclassReader</span><span class="p">,</span> <span class="n">NpBufferedWriter</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Open a file according to its suffix</span>

<span class="sd">    Open a `NpDataclassReader` file object, that can be used to read the file,</span>
<span class="sd">    either in chunks or completely. Files read in chunks can be used together with</span>
<span class="sd">    the `@bnp.streamable` decorator to call a function on all chunks in the file</span>
<span class="sd">    and optionally reduce the results.</span>

<span class="sd">    If `mode=&quot;w&quot;` it opens a writer object. </span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    filename : str</span>
<span class="sd">        Name of the file to open</span>
<span class="sd">    mode : str</span>
<span class="sd">        Either &quot;w&quot; or &quot;r&quot;</span>
<span class="sd">    buffer_type : FileBuffer</span>
<span class="sd">        A `FileBuffer` class to specify how the data in the file should be interpreted</span>
<span class="sd">    lazy : bool</span>
<span class="sd">        If True, the data will be read lazily, i. e. only when it is accessed. This is useful</span>
<span class="sd">        to speed up reading of large files, but it is more memory demanding</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    NpDataclassReader</span>
<span class="sd">        A file reader object</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; import bionumpy as bnp</span>
<span class="sd">    &gt;&gt;&gt; all_data = bnp.open(&quot;example_data/big.fq.gz&quot;).read()</span>
<span class="sd">    &gt;&gt;&gt; print(all_data)</span>
<span class="sd">    SequenceEntryWithQuality with 1000 entries</span>
<span class="sd">                         name                 sequence                  quality</span>
<span class="sd">      2fa9ee19-5c51-4281-a...  CGGTAGCCAGCTGCGTTCAG...  [10  5  5 12  5  4  3  </span>
<span class="sd">      1f9ca490-2f25-484a-8...  GATGCATACTTCGTTCGATT...  [ 5  4  5  4  6  6  5  </span>
<span class="sd">      06936a64-6c08-40e9-8...  GTTTTGTCGCTGCGTTCAGT...  [ 3  5  6  7  7  5  4  </span>
<span class="sd">      d6a555a1-d8dd-4e55-9...  CGTATGCTTTGAGATTCATT...  [ 2  3  4  4  4  4  6  </span>
<span class="sd">      91ca9c6c-12fe-4255-8...  CGGTGTACTTCGTTCCAGCT...  [ 4  3  5  6  3  5  6  </span>
<span class="sd">      4dbe5037-abe2-4176-8...  GCAGGTGATGCTTTGGTTCA...  [ 2  3  4  6  7  7  6  </span>
<span class="sd">      df3de4e9-48ca-45fc-8...  CATGCTTCGTTGGTTACCTC...  [ 5  5  5  4  7  7  7  </span>
<span class="sd">      bfde9b59-2f6d-48e8-8...  CTGTTGTGCGCTTCGTTCAT...  [ 8  8 10  7  8  6  3  </span>
<span class="sd">      dbcfd59a-7a96-46a2-9...  CGATTATTTGGTTCGTTCAT...  [ 5  4  2  3  5  2  2  </span>
<span class="sd">      a0f83c4e-4c20-4c15-b...  GTTGTACTTTACGTTTCAAT...  [ 3  5 10  6  7  6  6  </span>
<span class="sd">        </span>
<span class="sd">    &gt;&gt;&gt; first_chunk = bnp.open(&quot;example_data/big.fq.gz&quot;).read_chunk(300000)</span>
<span class="sd">    &gt;&gt;&gt; print(first_chunk)</span>
<span class="sd">    SequenceEntryWithQuality with 511 entries</span>
<span class="sd">                         name                 sequence                  quality</span>
<span class="sd">      2fa9ee19-5c51-4281-a...  CGGTAGCCAGCTGCGTTCAG...  [10  5  5 12  5  4  3  </span>
<span class="sd">      1f9ca490-2f25-484a-8...  GATGCATACTTCGTTCGATT...  [ 5  4  5  4  6  6  5  </span>
<span class="sd">      06936a64-6c08-40e9-8...  GTTTTGTCGCTGCGTTCAGT...  [ 3  5  6  7  7  5  4  </span>
<span class="sd">      d6a555a1-d8dd-4e55-9...  CGTATGCTTTGAGATTCATT...  [ 2  3  4  4  4  4  6  </span>
<span class="sd">      91ca9c6c-12fe-4255-8...  CGGTGTACTTCGTTCCAGCT...  [ 4  3  5  6  3  5  6  </span>
<span class="sd">      4dbe5037-abe2-4176-8...  GCAGGTGATGCTTTGGTTCA...  [ 2  3  4  6  7  7  6  </span>
<span class="sd">      df3de4e9-48ca-45fc-8...  CATGCTTCGTTGGTTACCTC...  [ 5  5  5  4  7  7  7  </span>
<span class="sd">      bfde9b59-2f6d-48e8-8...  CTGTTGTGCGCTTCGTTCAT...  [ 8  8 10  7  8  6  3  </span>
<span class="sd">      dbcfd59a-7a96-46a2-9...  CGATTATTTGGTTCGTTCAT...  [ 5  4  2  3  5  2  2  </span>
<span class="sd">      a0f83c4e-4c20-4c15-b...  GTTGTACTTTACGTTTCAAT...  [ 3  5 10  6  7  6  6  </span>
<span class="sd">    </span>
<span class="sd">    &gt;&gt;&gt; all_chunks = bnp.open(&quot;example_data/big.fq.gz&quot;).read_chunks(300000)</span>
<span class="sd">    </span>
<span class="sd">    &gt;&gt;&gt; for chunk in all_chunks:</span>
<span class="sd">    ...       print(chunk)</span>
<span class="sd">    ...</span>
<span class="sd">    SequenceEntryWithQuality with 511 entries</span>
<span class="sd">                         name                 sequence                  quality</span>
<span class="sd">      2fa9ee19-5c51-4281-a...  CGGTAGCCAGCTGCGTTCAG...  [10  5  5 12  5  4  3  </span>
<span class="sd">      1f9ca490-2f25-484a-8...  GATGCATACTTCGTTCGATT...  [ 5  4  5  4  6  6  5  </span>
<span class="sd">      06936a64-6c08-40e9-8...  GTTTTGTCGCTGCGTTCAGT...  [ 3  5  6  7  7  5  4  </span>
<span class="sd">      d6a555a1-d8dd-4e55-9...  CGTATGCTTTGAGATTCATT...  [ 2  3  4  4  4  4  6  </span>
<span class="sd">      91ca9c6c-12fe-4255-8...  CGGTGTACTTCGTTCCAGCT...  [ 4  3  5  6  3  5  6  </span>
<span class="sd">      4dbe5037-abe2-4176-8...  GCAGGTGATGCTTTGGTTCA...  [ 2  3  4  6  7  7  6  </span>
<span class="sd">      df3de4e9-48ca-45fc-8...  CATGCTTCGTTGGTTACCTC...  [ 5  5  5  4  7  7  7  </span>
<span class="sd">      bfde9b59-2f6d-48e8-8...  CTGTTGTGCGCTTCGTTCAT...  [ 8  8 10  7  8  6  3  </span>
<span class="sd">      dbcfd59a-7a96-46a2-9...  CGATTATTTGGTTCGTTCAT...  [ 5  4  2  3  5  2  2  </span>
<span class="sd">      a0f83c4e-4c20-4c15-b...  GTTGTACTTTACGTTTCAAT...  [ 3  5 10  6  7  6  6  </span>
<span class="sd">    SequenceEntryWithQuality with 489 entries</span>
<span class="sd">                         name                 sequence                  quality</span>
<span class="sd">      5f27fb90-2cb0-43d0-a...  CGTTGCTGATTCAGCATCAA...  [ 5  3  2  3  2  2  4  </span>
<span class="sd">      e23294d9-0079-4345-a...  CGAGCCGCTTCGTTCCGGTT...  [ 4  5  3  3  3  4  3  </span>
<span class="sd">      56736851-ccc9-41a6-9...  CGGTGCCTTCGTTCATTTCT...  [ 8  3  7  7  3  1  2  </span>
<span class="sd">      f156362d-d380-480d-8...  CTGTTGCGCCCCGGAACAGT...  [ 7 11  9  4  4  4  3  </span>
<span class="sd">      300f89ef-608a-463f-8...  CATACTTTGGTTCATTCTGT...  [ 3  2  4  4  4  4  5  </span>
<span class="sd">      755b1702-4560-4c04-a...  GGTATACTTGCCCTACGTTC...  [10  9 13  6  3  3  4  </span>
<span class="sd">      98de4f6b-d094-41e8-9...  GTTGTACTTCGTTCAGTTTC...  [ 4  5  6  4  7  6  6  </span>
<span class="sd">      00ac3f41-f735-49e5-9...  GTTGTACTTCGTTCAGCTCT...  [ 3  4  5  4  4 10 12 1</span>
<span class="sd">      f92d30bc-f77f-401e-9...  GTTGTACTGCTTCGTTCAGT...  [ 6  3  4  3  6  3  2  </span>
<span class="sd">      7e2c14c0-0662-4cc3-8...  TGATACATTACTTCGTTCGA...  [ 3  8  4  7  2  4  3  </span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">path</span> <span class="o">=</span> <span class="n">PurePath</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">suffix</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">suffixes</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">is_gzip</span> <span class="o">=</span> <span class="n">suffix</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;.bam&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.gz&quot;</span><span class="p">:</span>
        <span class="n">suffix</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">suffixes</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">_get_buffered_file</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">suffix</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">is_gzip</span><span class="o">=</span><span class="n">is_gzip</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">buffer_type</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="n">lazy</span><span class="p">)</span></div>


    
<div class="viewcode-block" id="count_entries">
<a class="viewcode-back" href="../../../modules/io.html#bionumpy.io.count_entries">[docs]</a>
<span class="k">def</span> <span class="nf">count_entries</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">buffer_type</span><span class="p">:</span> <span class="n">FileBuffer</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Count the number of entries in the file</span>

<span class="sd">    By default it uses the file suffix to imply the file format. But</span>
<span class="sd">    a specific `FileBuffer` can be provided.</span>


<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    filename : str</span>
<span class="sd">        Name of the file to count the entries of</span>
<span class="sd">    buffer_type : FileBuffer</span>
<span class="sd">        A `FileBuffer` class to specify how the data in the file should be interpreted</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        The number of entries in the file</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    6</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">NumpyFileReader</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Counting entries in </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">PurePath</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">suffix</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">suffixes</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">is_gzip</span> <span class="o">=</span> <span class="n">suffix</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;.bam&quot;</span><span class="p">)</span>
    <span class="c1">#if suffix == &#39;.bam&#39;:</span>
    <span class="c1">#    reader = NumpyBamReader</span>
    <span class="k">if</span> <span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.gz&quot;</span><span class="p">:</span>
        <span class="n">suffix</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">suffixes</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">open_func</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span> <span class="k">if</span> <span class="n">is_gzip</span> <span class="k">else</span> <span class="nb">open</span>
    <span class="k">if</span> <span class="n">buffer_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">buffer_type</span> <span class="o">=</span> <span class="n">_get_buffer_type</span><span class="p">(</span><span class="n">suffix</span><span class="p">)</span>


    <span class="n">file_reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">(</span><span class="n">open_func</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">),</span> <span class="n">buffer_type</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">is_gzip</span><span class="p">:</span>
        <span class="n">file_reader</span><span class="o">.</span><span class="n">set_prepend_mode</span><span class="p">()</span>
    <span class="n">chunk_counts</span> <span class="o">=</span> <span class="p">(</span><span class="n">chunk</span><span class="o">.</span><span class="n">count_entries</span><span class="p">()</span> <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">file_reader</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">(</span><span class="n">min_chunk_size</span><span class="o">=</span><span class="mi">500000</span><span class="p">))</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">chunk_counts</span><span class="p">)</span></div>



<span class="k">def</span> <span class="nf">read</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">mode</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">buffer_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">FileBuffer</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BNPDataClass</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read the content of a file</span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    filename: str</span>
<span class="sd">    mode: str</span>
<span class="sd">    buffer_type:</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    BNPDataClass</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">bnp_open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">buffer_type</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">content</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">content</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>