

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bionumpy.arithmetics.similarity_measures &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../menu.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bionumpy.arithmetics.similarity_measures</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bionumpy.arithmetics.similarity_measures</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">.intervals</span> <span class="kn">import</span> <span class="n">get_boolean_mask</span>
<span class="kn">from</span> <span class="nn">..streams</span> <span class="kn">import</span> <span class="n">MultiStream</span><span class="p">,</span> <span class="n">streamable</span><span class="p">,</span> <span class="n">BnpStream</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">..datatypes</span> <span class="kn">import</span> <span class="n">ChromosomeSize</span><span class="p">,</span> <span class="n">Interval</span>


<span class="nd">@streamable</span><span class="p">(</span><span class="nb">sum</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">get_contingency_table</span><span class="p">(</span><span class="n">intervals_a</span><span class="p">,</span> <span class="n">intervals_b</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">):</span>
    <span class="n">boolean_a</span> <span class="o">=</span> <span class="n">get_boolean_mask</span><span class="p">(</span><span class="n">intervals_a</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">)</span>
    <span class="n">not_a</span> <span class="o">=</span> <span class="o">~</span><span class="n">boolean_a</span>
    <span class="n">boolean_b</span> <span class="o">=</span> <span class="n">get_boolean_mask</span><span class="p">(</span><span class="n">intervals_b</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">)</span>
    <span class="n">not_b</span> <span class="o">=</span> <span class="o">~</span><span class="n">boolean_b</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">boolean_a</span> <span class="o">&amp;</span> <span class="n">boolean_b</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">boolean_a</span> <span class="o">&amp;</span> <span class="n">not_b</span><span class="p">)],</span>
                     <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">not_a</span> <span class="o">&amp;</span> <span class="n">boolean_b</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">not_a</span> <span class="o">&amp;</span> <span class="n">not_b</span><span class="p">)]])</span>


<div class="viewcode-block" id="forbes">
<a class="viewcode-back" href="../../../modules/genome_arithmetics.html#bionumpy.arithmetics.forbes">[docs]</a>
<span class="k">def</span> <span class="nf">forbes</span><span class="p">(</span><span class="n">chromosome_sizes</span><span class="p">:</span> <span class="n">ChromosomeSize</span><span class="p">,</span> <span class="n">intervals_a</span><span class="p">:</span> <span class="n">Interval</span><span class="p">,</span> <span class="n">intervals_b</span><span class="p">:</span> <span class="n">Interval</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Computes the Forbes similarity index for two sets of intervals.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    chromosome_sizes : ChromosomeSize</span>
<span class="sd">        A ChromosomeSizes, typically from reading a chromosome.sizes file with bnp.open()</span>
<span class="sd">    intervals_a : Interval</span>
<span class="sd">        Must be sorted. Can be read using bnp.open on a bed-file.</span>
<span class="sd">    intervals_a : Interval</span>
<span class="sd">        Must be sorted. Can be read using bnp.open on a bed-file.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        The forbes similarity index.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; from bionumpy.arithmetics import forbes, sort_intervals</span>
<span class="sd">    &gt;&gt;&gt; from bionumpy.datatypes import Interval</span>
<span class="sd">    &gt;&gt;&gt; a = Interval.from_entry_tuples([(&quot;chr1&quot;, 10, 20), (&quot;chr2&quot;, 20, 30)])</span>
<span class="sd">    &gt;&gt;&gt; b = Interval.from_entry_tuples([(&quot;chr2&quot;, 15, 25), (&quot;chr1&quot;, 10, 40)])</span>
<span class="sd">    &gt;&gt;&gt; a_sorted = sort_intervals(a, sort_order=[&quot;chr1&quot;, &quot;chr2&quot;])</span>
<span class="sd">    &gt;&gt;&gt; b_sorted = sort_intervals(b, sort_order=[&quot;chr1&quot;, &quot;chr2&quot;])</span>
<span class="sd">    &gt;&gt;&gt; forbes({&quot;chr1&quot;: 100, &quot;chr2&quot;: 200}, a_sorted, b_sorted)</span>
<span class="sd">    5.625</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ms</span> <span class="o">=</span> <span class="n">MultiStream</span><span class="p">(</span><span class="n">chromosome_sizes</span><span class="p">,</span> <span class="n">a</span><span class="o">=</span><span class="n">intervals_a</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="n">intervals_b</span><span class="p">)</span>
    <span class="p">((</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span> <span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span> <span class="o">=</span> <span class="n">get_contingency_table</span><span class="p">(</span><span class="n">ms</span><span class="o">.</span><span class="n">a</span><span class="p">,</span> <span class="n">ms</span><span class="o">.</span><span class="n">b</span><span class="p">,</span> <span class="n">ms</span><span class="o">.</span><span class="n">lengths</span><span class="p">)</span>
    <span class="n">N</span> <span class="o">=</span> <span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">a</span><span class="o">*</span><span class="n">N</span><span class="o">/</span><span class="p">((</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">c</span><span class="p">)))</span></div>



<div class="viewcode-block" id="jaccard">
<a class="viewcode-back" href="../../../modules/genome_arithmetics.html#bionumpy.arithmetics.jaccard">[docs]</a>
<span class="k">def</span> <span class="nf">jaccard</span><span class="p">(</span><span class="n">chromosome_sizes</span><span class="p">:</span> <span class="n">ChromosomeSize</span><span class="p">,</span> <span class="n">intervals_a</span><span class="p">:</span> <span class="n">Interval</span><span class="p">,</span> <span class="n">intervals_b</span><span class="p">:</span> <span class="n">Interval</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Computes the Jaccard similarity index for two sets of intervals.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    chromosome_sizes : ChromosomeSize</span>
<span class="sd">        A ChromosomeSizes, typically from reading a chromosome.sizes file with bnp.open()</span>
<span class="sd">    intervals_a : Interval</span>
<span class="sd">        Must be sorted. Can be read using bnp.open on a bed-file.</span>
<span class="sd">    intervals_a : Interval</span>
<span class="sd">        Must be sorted. Can be read using bnp.open on a bed-file.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        The forbes similarity index.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    See forbes for examples.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ms</span> <span class="o">=</span> <span class="n">MultiStream</span><span class="p">(</span><span class="n">chromosome_sizes</span><span class="p">,</span> <span class="n">a</span><span class="o">=</span><span class="n">intervals_a</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="n">intervals_b</span><span class="p">)</span>
    <span class="p">((</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span> <span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span> <span class="o">=</span> <span class="n">get_contingency_table</span><span class="p">(</span><span class="n">ms</span><span class="o">.</span><span class="n">a</span><span class="p">,</span> <span class="n">ms</span><span class="o">.</span><span class="n">b</span><span class="p">,</span> <span class="n">ms</span><span class="o">.</span><span class="n">lengths</span><span class="p">)</span>
    <span class="n">N</span> <span class="o">=</span> <span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">a</span><span class="o">/</span><span class="p">(</span><span class="n">N</span><span class="o">-</span><span class="n">d</span><span class="p">))</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>