

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Genomic Data &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Genomic Data</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/topics/genomic_data.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="genomic-data">
<h1>Genomic Data<a class="headerlink" href="#genomic-data" title="Link to this heading"></a></h1>
<p>Analysis of genomic data is best done with the knowledge of how the genome looks like. At the very least, we should know how the names of each chromosome in the genome, and how long the sequence of each chromosome is. This serves several purposes:</p>
<ul class="simple">
<li><p>We (and the computer) become aware of which reference genome we are using, which can prevent errors from mismatched reference genomes</p></li>
<li><p>Chromosomes without any data gets handled correctly</p></li>
<li><p>We get a speed increase in the algorithms</p></li>
</ul>
<p>For these reasons, we always recomend using the genomic data framework for handling genomic data in BioNumPy, not using raw bed, vcf or bdg files directly from the <cite>bnp.open</cite> interface. The main access point of the genomic data interface is the genome object, which can be created directly from a <cite>dict</cite> of chromosome names and sizes; or, more commonly, from a <cite>.chrom.sizes</cite> or <cite>.fa.fai</cite> file, which are tab separated lines of chromosome names and sizes:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="n">genome</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">Genome</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="s2">&quot;example_data/small.chrom.sizes&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Chromosome                     Size
      chr1                    10000
      chr2                    20000
      chr3                    30000
</pre></div>
</div>
<p>From this genome object we can create genomic data from reading files, or from bnpdataclass objects directly. For instance, we can now read a set of intervals from a <cite>.bed</cite> file and get a <cite>GenomicIntervals</cite> object. Because  <cite>GenomicIntervals</cite> objects knows which reference genome they belong to, they are easier and faster to work with than raw ‘bnp.Interval’ objects.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">intervals</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_intervals</span><span class="p">(</span><span class="s1">&#39;example_data/small_peaks.narrowPeak&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">intervals</span><span class="p">)</span>
<span class="n">mask</span> <span class="o">=</span> <span class="n">intervals</span><span class="o">.</span><span class="n">get_mask</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> Genomic Intervals on [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;]:
Interval with 13 entries
                chromosome                    start                     stop
                      chr1                      482                      790
                      chr1                     5873                     6183
                      chr1                     6969                     7289
                      chr2                      691                     1007
                      chr2                     6155                     6459
                      chr2                     8337                     8641
                      chr2                    11170                    11481
                      chr2                    12375                    12677
                      chr2                    12918                    13232
                      chr2                    18804                    19134
 chr1: [ False False False ... False False False]
 chr2: [ False False False ... False False False]
 chr3: [ False False False ... False False False]
</pre></div>
</div>
<p>The mask here is a boolean array which is <cite>True</cite> wherever any of the intervals overlap. We can also get a <cite>GenomicArray</cite> from a bedgraph file by using <cite>genome.read_track()</cite></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pileup</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_track</span><span class="p">(</span><span class="s1">&#39;example_data/small_treat_pileup.bdg&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pileup</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chr1: [ 0.0 0.0 1.0 ... 0.0 0.0 0.0]
chr2: [ 0.0 0.0 0.0 ... 0.0 0.0 0.0]
chr3: [ 0.0 0.0 0.0 ... 0.0 0.0 0.0]
</pre></div>
</div>
<p>A benefit of working with genomic data in this way is that they cooperate in a consistent way. We can use GenomicIntervals as indexes to GenomicData, and can again use the extracted data to filter the intervals. Let’s see how the <cite>treat_pileup</cite> looks in peak areas (we get the max and mean pileup value for each peak):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">peak_pileups</span> <span class="o">=</span> <span class="n">pileup</span><span class="p">[</span><span class="n">intervals</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">peak_pileups</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">peak_pileups</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[ 227.  231.  412.  296.  165.  163.  271.  148.  268.  568. 1901.   90.
  236.]
[145.68181818 147.86129032 256.671875   187.71202532 106.79934211
 106.23355263 173.04501608  96.27152318 169.59235669 346.37272727
 593.44084507  60.25614035 150.72580645]
</pre></div>
</div>
<p>we can now again use these values to filter the intervals based on the treat_pileup values in each interval. For instance only keep peaks with a max treatment_value above 200:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">high_pileup_peaks</span> <span class="o">=</span> <span class="n">intervals</span><span class="p">[</span><span class="n">peak_pileups</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">200</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">high_pileup_peaks</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Genomic Intervals on [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;]:
 Interval with 9 entries
                chromosome                    start                     stop
                      chr1                      482                      790
                      chr1                     5873                     6183
                      chr1                     6969                     7289
                      chr2                      691                     1007
                      chr2                    11170                    11481
                      chr2                    12918                    13232
                      chr2                    18804                    19134
                      chr3                    10677                    11387
                      chr3                    27057                    27367
</pre></div>
</div>
<p>A further way to analyze these peaks is to check the sequence in the peaks for motifs. We can load the reference sequence using the <cite>genome.read_sequence()</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">genome_sequence</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">read_sequence</span><span class="p">(</span><span class="s1">&#39;example_data/small_sequence.fa&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">genome_sequence</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>GenomicSequence over chromosomes: [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;]
</pre></div>
</div>
<p>Now we can use our intervals as indexed to the reference sequence in much the same way as with genomic arrays. This we can use to get the sequences of the peaks and check them for motifs:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span> <span class="n">peak_sequences</span> <span class="o">=</span> <span class="n">genome_sequence</span><span class="p">[</span><span class="n">high_pileup_peaks</span><span class="p">]</span>
 <span class="nb">print</span><span class="p">(</span><span class="n">peak_sequences</span><span class="p">)</span>
 <span class="kn">from</span> <span class="nn">pyjaspar</span> <span class="kn">import</span> <span class="n">jaspardb</span>
 <span class="kn">from</span> <span class="nn">bionumpy.sequence</span> <span class="kn">import</span> <span class="n">PWM</span>
 <span class="n">jaspar_object</span> <span class="o">=</span> <span class="n">jaspardb</span><span class="p">(</span><span class="n">release</span><span class="o">=</span><span class="s2">&quot;JASPAR2020&quot;</span><span class="p">)</span>
 <span class="n">ctcf_motif</span> <span class="o">=</span> <span class="n">jaspar_object</span><span class="o">.</span><span class="n">fetch_motifs_by_name</span><span class="p">(</span><span class="s1">&#39;CTCF&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
 <span class="n">motif</span> <span class="o">=</span> <span class="n">PWM</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">ctcf_motif</span><span class="o">.</span><span class="n">pwm</span><span class="p">)</span>
 <span class="nb">print</span><span class="p">(</span><span class="n">bnp</span><span class="o">.</span><span class="n">get_motif_scores</span><span class="p">(</span><span class="n">peak_sequences</span><span class="p">,</span> <span class="n">motif</span><span class="p">)</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> AGAGCCGGACCGAATGACGT...
 TCAGGTAGAACTCGCATTTC...
 AAGACTTATTTGATGGCCGG...
 ATAGAGAGCGTTCGGCGCTA...
 CTCCTTAGCATACAAACGGG...
 TGCCCCATCTCTACACAATT...
 GCGCTGCCGTCACGGCGGGG...
 TTACATCCTGACGAAATACA...
 GACGGGTAGGCGATTTTTAT...
 [-0.82852725 -4.78231564 -0.90188366 -4.97871502 -3.06703885 -1.00359223
   7.75920829  0.19382436  2.5188205 ]
</pre></div>
</div>
<section id="further-reading">
<h2>Further reading<a class="headerlink" href="#further-reading" title="Link to this heading"></a></h2>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="../tutorials/genomic_data.html#multi-genomic-data"><span class="std std-ref">Tutorail: Genomic data from multiple sources</span></a></p></li>
<li><p><a class="reference internal" href="../tutorials/subsetting_bed.html#subsetting-bed"><span class="std std-ref">Tutorial: Analysing the read pileup within peaks (intervals)</span></a></p></li>
<li><p><a class="reference internal" href="../tutorials/similarity_measures.html#similarity-measures-tutorial"><span class="std std-ref">Tutorial: Computing the similarity between to bed files</span></a></p></li>
<li><p><a class="reference internal" href="../source/intervals.html#intervals"><span class="std std-ref">More about intervals</span></a></p></li>
<li><p><a class="reference internal" href="../modules/genomic_data.html#arithmetics-api"><span class="std std-ref">API documentation on the arithmetics module</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>