

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Using BioNumPy to analyse kmers, minimzers and motifs &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_bionumpy_in_your_existing_project.html">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="../interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Using BioNumPy to analyse kmers, minimzers and motifs</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/topics/kmers.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="using-bionumpy-to-analyse-kmers-minimzers-and-motifs">
<h1>Using BioNumPy to analyse kmers, minimzers and motifs<a class="headerlink" href="#using-bionumpy-to-analyse-kmers-minimzers-and-motifs" title="Link to this heading"></a></h1>
<p>BioNumPy has functionality for easily finding kmers and minimizers in a set of sequences, and does so very efficiently.</p>
<p>For instance, finding all 3-mers and counting the number of <cite>ACT</cite> in some sequences is as simple as:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">as_encoded_array</span><span class="p">([</span><span class="s2">&quot;ACTG&quot;</span><span class="p">,</span> <span class="s2">&quot;GGGACT&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">],</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kmers</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">get_kmers</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kmers</span>
<span class="go">encoded_ragged_array([[ACT, CTG],</span>
<span class="go">                      [GGG, GGA, GAC, ACT],</span>
<span class="go">                      []], 3merEncoding(AlphabetEncoding(&#39;ACGT&#39;)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">count_encoded</span><span class="p">(</span><span class="n">kmers</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;ACT&quot;</span><span class="p">]</span>
<span class="go">array(2)</span>
</pre></div>
</div>
<p>The same functionality works for minimizers:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bnp</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">get_minimizers</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">encoded_ragged_array([[AC],</span>
<span class="go">                      [GA, GA, GA],</span>
<span class="go">                      []], 2merEncoding(AlphabetEncoding(&#39;ACGT&#39;)))</span>
</pre></div>
</div>
<p>Note that the kmer hashes and minimizers are stored numerically internally, and how they are computed will depend on the encoding of the sequences.
If your sequences is encoded with an encoding with alphabet size <strong>4</strong> (e.g. <cite>bnp.DNAEncoding</cite>)
<cite>get_kmers</cite> is optimized to run a lot faster by performing bit-operations.</p>
<section id="analysing-kmers-and-minimizers-in-big-datasets">
<h2>Analysing kmers and minimizers in big datasets<a class="headerlink" href="#analysing-kmers-and-minimizers-in-big-datasets" title="Link to this heading"></a></h2>
<p>BioNumPy is able to quite efficiently find all kmers or minimizers in large sequence datasets. Since the resulting kmers are stored numerically in NumPy arrays, NumPy functionality can be used to do analysis.</p>
<p>Below in an example of finding all 31-mers in a fastq file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">file</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/big.fq.gz&quot;</span><span class="p">)</span>
<span class="c1"># read the file chunk by chunk to keep memory low:</span>
<span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="n">chunk</span><span class="o">.</span><span class="n">sequence</span>
    <span class="c1"># change encoding to a DNAEncoding, this works as long as the</span>
    <span class="c1"># sequences only contains ACGT, and makes get_kmers extremely efficient</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">change_encoding</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Kmers:&quot;</span><span class="p">)</span>
    <span class="n">kmers</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">get_kmers</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">31</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">kmers</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">])</span>

    <span class="c1"># kmers is an EncodedRaggedArray, one row for each read</span>
    <span class="c1"># you can get all the raw numeric kmers efficiently like this:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Raw kmers:&quot;</span><span class="p">)</span>
    <span class="n">numeric_kmers</span> <span class="o">=</span> <span class="n">kmers</span><span class="o">.</span><span class="n">raw</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">numeric_kmers</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">])</span>

    <span class="c1"># and if you don&#39;t care about the RaggedStructure, you can do ravel:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Flat raw kmers:&quot;</span><span class="p">)</span>
    <span class="n">numeric_flat_kmers</span> <span class="o">=</span> <span class="n">numeric_kmers</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">numeric_flat_kmers</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
<p>Output from the code above:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Kmers:
[CGGTAGCCAGCTGCGTTCAGTATGGAAGATT, GGTAGCCAGCTGCGTTCAGTATGGAAGATTT]
[GATGCATACTTCGTTCGATTTCGTTTCAACT, ATGCATACTTCGTTCGATTTCGTTTCAACTG]
[GTTTTGTCGCTGCGTTCAGTTTATGGGTGCG, TTTTGTCGCTGCGTTCAGTTTATGGGTGCGG]
Raw kmers:
[4360244785522956521 4548825710201280058]
[3755975642940518834 3244836919948823660]
[2804282287455632382 3006913581077602047]
Flat raw kmers:
[4360244785522956521 4548825710201280058 3443049436764013966
  860762359191003491]
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>