

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Using BioNumPy in your existing project &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Filtering FASTQ reads" href="tutorials/fastq_filtering.html" />
    <link rel="prev" title="Getting started with BioNumPy in 10 minutes" href="introduction.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="menu.html" class="icon icon-home">
            bionumpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">BioNumPy at a glance</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Getting started with BioNumPy in 10 minutes</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Using BioNumPy in your existing project</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/subsetting_bed.html">Getting read pileup inside peaks (intervals)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/extracting_kmers_around_snps.html">Extracting kmers around snps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Interoperability</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="interoperability/numpy_operations.html">Numpy Interoperability</a></li>
<li class="toctree-l1"><a class="reference internal" href="interoperability/working_with_biopython.html">Using BioNumPy with Biopython</a></li>
<li class="toctree-l1"><a class="reference internal" href="interoperability/pandas.html">Pandas Interoperability</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="source/supported_file_formats.html#reading-a-new-file-format">Reading a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/genomic_data.html">Genomic Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="menu.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Using BioNumPy in your existing project</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/using_bionumpy_in_your_existing_project.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="using-bionumpy-in-your-existing-project">
<span id="id1"></span><h1>Using BioNumPy in your existing project<a class="headerlink" href="#using-bionumpy-in-your-existing-project" title="Link to this heading"></a></h1>
<p>If you already have a Python project that reads biological datasets (e.g. fastq, bam or vcf files), you can often easily add BioNumPy to your project to speed up reading and parsing of such files.</p>
<p>All you need is to install BioNumPy, import it, and read your file with <cite>bnp.open</cite>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>

<span class="c1"># open your file, bnp.open automatically detects the file format</span>
<span class="n">f</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/variants.vcf&quot;</span><span class="p">)</span>
    <span class="c1"># a chunk is an efficient representation of a chunk of many lines</span>
<span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
            <span class="c1"># we can iterate over the entries</span>
    <span class="k">for</span> <span class="n">single_entry</span> <span class="ow">in</span> <span class="n">chunk</span><span class="o">.</span><span class="n">to_iter</span><span class="p">():</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">single_entry</span><span class="p">)</span>
                    <span class="c1"># and we can access things like, chromosome, position and so on</span>
                    <span class="n">position</span> <span class="o">=</span> <span class="n">single_entry</span><span class="o">.</span><span class="n">position</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>VCFEntry(chromosome=&#39;chr1&#39;, position=883624, id=&#39;rs4970378&#39;, ref_seq=&#39;A&#39;, alt_seq=&#39;G&#39;, quality=&#39;.&#39;, filter=&#39;.&#39;, info=&#39;.&#39;)
VCFEntry(chromosome=&#39;chr1&#39;, position=887559, id=&#39;rs3748595&#39;, ref_seq=&#39;A&#39;, alt_seq=&#39;C&#39;, quality=&#39;.&#39;, filter=&#39;.&#39;, info=&#39;.&#39;)
VCFEntry(chromosome=&#39;chr1&#39;, position=887800, id=&#39;rs3828047&#39;, ref_seq=&#39;A&#39;, alt_seq=&#39;G&#39;, quality=&#39;.&#39;, filter=&#39;.&#39;, info=&#39;.&#39;)
</pre></div>
</div>
<p>The above example shows how you can read any file and iterate over the lines. This way, one can in most cases easily use BioNumPy as an inplace replacement for the existing code that iterates over lines from biological data files. However, iterating over lines is not very efficient, so we can do better by working directly on chunks of the file. See <a class="reference internal" href="source/reading_files.html#reading-files"><span class="std std-ref">Reading files</span></a> for more information.</p>
<section id="using-bionumpy-with-pandas">
<h2>Using BioNumPy with Pandas<a class="headerlink" href="#using-bionumpy-with-pandas" title="Link to this heading"></a></h2>
<p>Another way that BioNumPy can be used directly in existing projects is if you are using Pandas.</p>
<p>It is straightforward to convert a chunk of a file or a whole file to a Pandas DataFrame. This time, we read a fastq file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="n">f</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/big.fq.gz&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">():</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">chunk</span><span class="o">.</span><span class="n">topandas</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>                              name                       sequence                        quality
0    2fa9ee19-5c51-4281-abdd-ea...  CGGTAGCCAGCTGCGTTCAGTATGGA...  [10, 5, 5, 12, 5, 4, 3, 4,...
1    1f9ca490-2f25-484a-8972-d6...  GATGCATACTTCGTTCGATTTCGTTT...  [5, 4, 5, 4, 6, 6, 5, 6, 1...
2    06936a64-6c08-40e9-8a10-0f...  GTTTTGTCGCTGCGTTCAGTTTATGG...  [3, 5, 6, 7, 7, 5, 4, 3, 3...
3    d6a555a1-d8dd-4e55-936f-ad...  CGTATGCTTTGAGATTCATTCAGGAG...  [2, 3, 4, 4, 4, 4, 6, 6, 7...
4    91ca9c6c-12fe-4255-83cc-96...  CGGTGTACTTCGTTCCAGCTAGATTT...  [4, 3, 5, 6, 3, 5, 6, 5, 5...
..                             ...                            ...                            ...
995  2eef382a-21f7-4a5b-a8d8-64...  CGTTTGCGCTGGTTCATTTTATCGGT...  [2, 6, 2, 2, 4, 2, 3, 3, 4...
996  18949e40-d30d-49f7-8a1c-c2...  GCGTACTTCGTTCAGTTTCGGAAGTG...  [2, 2, 2, 3, 3, 3, 5, 7, 7...
997  f4aeadf5-174e-4974-aef1-8b...  CAGTAATACTTCGTTCCAGTTCTGGG...  [9, 6, 11, 10, 2, 3, 3, 3,...
998  6b3cb23e-3f71-435b-835f-78...  CTGTTGTACTTCGATTCATTCAGGTG...  [5, 3, 3, 5, 6, 4, 5, 3, 5...
999  d65b5418-65d5-4bf3-aac8-aa...  CGGTGACGCTGGTTTAAATCTAACGG...  [7, 3, 4, 3, 4, 2, 2, 3, 3...

[1000 rows x 3 columns]
</pre></div>
</div>
<p>Continue to see an overview of <a class="reference internal" href="index.html#what-can-you-do"><span class="std std-ref">what you can do with bionumpy</span></a>.</p>
</section>
<section id="using-bionumpy-raggedarray-functionalities">
<h2>Using BioNumPy RaggedArray functionalities<a class="headerlink" href="#using-bionumpy-raggedarray-functionalities" title="Link to this heading"></a></h2>
<p>Alternatively, if one is working with biological sequences and are already loaded into a list of strings,
one can use the <cite>as_encoded_array</cite> function to convert the list of strings into a RaggedArray.
This is useful if you want to make use of all the efficient ways of BioNumpy’s functionalities on
handling non-uniform length sequences and computations on them. Below is an example. See all the available encodings in <a class="reference internal" href="source/encoding.html#encodings"><span class="std std-ref">Encodings</span></a>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="n">my_sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;TGTGCCAGCAGCGGGGATCGTAATCAGCCCCAGCATTTT&quot;</span><span class="p">,</span>
                <span class="s2">&quot;TGCAGCGTCAAGGTCCAAGCTTTCTTT&quot;</span><span class="p">,</span>
                <span class="s2">&quot;TGTGCCACCAGTGATTATTATTGGTACGAGCAGTACTTC&quot;</span><span class="p">]</span>
<span class="n">my_ragged_array</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">as_encoded_array</span><span class="p">(</span><span class="n">my_sequences</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">encodings</span><span class="o">.</span><span class="n">alphabet_encoding</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">my_ragged_array</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">my_ragged_array</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>TGTGCCAGCAGCGGGGATCGTAATCAGCCCCAGCATTTT
TGCAGCGTCAAGGTCCAAGCTTTCTTT
TGTGCCACCAGTGATTATTATTGGTACGAGCAGTACTTC
(3, array([39, 27, 39]))
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="introduction.html" class="btn btn-neutral float-left" title="Getting started with BioNumPy in 10 minutes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorials/fastq_filtering.html" class="btn btn-neutral float-right" title="Filtering FASTQ reads" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>